// @hubspot/ui-extensions/dist/shared/remoteComponents.js
import { createRemoteComponentRegistry } from "./utils/remote-component-registry.js";
/**
 * Represents a registry of HubSpot-provided React components that should only be used **internally** by the UI extension SDK.
 *
 * @internal
 */
export const __hubSpotComponentRegistry = createRemoteComponentRegistry();
const { createAndRegisterRemoteReactComponent, createAndRegisterRemoteCompoundReactComponent, } = __hubSpotComponentRegistry;
////////////////////////////////////////////////////////////
// STANDARD COMPONENTS
////////////////////////////////////////////////////////////
/**
 * The `Alert` component renders an alert within a card. Use this component to give usage guidance, notify users of action results, or warn them about potential issues or failures.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/alert Docs}
 * - {@link https://app.hubspot.com/docs/48008916/reference/ui-components/standard-components/alert#variants Variants}
 */
export const Alert = createAndRegisterRemoteReactComponent('Alert');
/**
 * The `Button` component renders a single button. Use this component to enable users to perform actions, such as submitting a form, sending data to an external system, or deleting data.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/button Docs}
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/button#usage-examples Examples}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/design-patterns#button Design Pattern Examples}
 */
export const Button = createAndRegisterRemoteReactComponent('Button', {
    fragmentProps: ['overlay'],
});
/**
 * The `ButtonRow` component renders a row of specified `Button` components. Use this component when you want to include multiple buttons in a row.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/button-row Docs}
 */
export const ButtonRow = createAndRegisterRemoteReactComponent('ButtonRow');
export const Card = createAndRegisterRemoteReactComponent('Card');
/**
 * The `DescriptionList` component renders pairs of labels and values. Use this component to display pairs of labels and values in a way that's easy to read at a glance.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/description-list Docs}
 */
export const DescriptionList = createAndRegisterRemoteReactComponent('DescriptionList');
/**
 * The `DescriptionListItem` component renders a single set of a label and value. Use this component within a `DescriptionList` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/description-list Docs}
 */
export const DescriptionListItem = createAndRegisterRemoteReactComponent('DescriptionListItem');
/**
 * The `Divider` component renders a grey, horizontal line for spacing out components vertically or creating sections in an extension. Use this component to space out other components when the content needs more separation than white space.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/divider Docs}
 */
export const Divider = createAndRegisterRemoteReactComponent('Divider');
/**
 * The `Spacer` component renders vertical space between components. Use this component
 * to add consistent spacing without using empty wrapper components.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/spacer Docs}
 */
export const Spacer = createAndRegisterRemoteReactComponent('Spacer');
/**
 * The `EmptyState` component sets the content that appears when the extension is in an empty state. Use this component when there's no content or data to help guide users.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/empty-state Docs}
 */
export const EmptyState = createAndRegisterRemoteReactComponent('EmptyState');
/**
 * The `ErrorState` component sets the content of an erroring extension. Use this component to guide users through resolving errors that your extension might encounter.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/error-state Docs}
 */
export const ErrorState = createAndRegisterRemoteReactComponent('ErrorState');
/**
 * The `Form` component renders a form that can contain other subcomponents, such as `Input`, `Select`, and `Button`. Use this component to enable users to submit data to HubSpot or an external system.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/form Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/design-patterns#form Design Pattern Examples}
 */
export const Form = createAndRegisterRemoteReactComponent('Form');
/**
 * The `Heading` component renders large heading text. Use this component to introduce or differentiate sections of your component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/heading Docs}
 */
export const Heading = createAndRegisterRemoteReactComponent('Heading');
/**
 * The `Image` component renders an image. Use this component to add a logo or other visual brand identity asset, or to accentuate other content in the extension.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/image Docs}
 */
export const Image = createAndRegisterRemoteReactComponent('Image', {
    fragmentProps: ['overlay'],
});
/**
 * The `Input` component renders a text input field where a user can enter a custom text value. Like other inputs, this component should be used within a `Form` that has a submit button.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/input Docs}
 */
export const Input = createAndRegisterRemoteReactComponent('Input');
/**
 * The `Link` component renders a clickable hyperlink. Use links to direct users to an external web page or another part of the HubSpot app.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/link Docs}
 */
export const Link = createAndRegisterRemoteReactComponent('Link', {
    fragmentProps: ['overlay'],
});
/**
 * The `TextArea` component renders a fillable text field. Like other inputs, this component should be used within a `Form` that has a submit button.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/text-area Docs}
 */
export const TextArea = createAndRegisterRemoteReactComponent('TextArea');
// Textarea was changed to TextArea
// Exporting both for backwards compat
/** @deprecated use TextArea instead. With a capital A.*/
export const Textarea = createAndRegisterRemoteReactComponent('Textarea');
/**
 * The `LoadingSpinner` component renders a visual indicator for when an extension is loading or processing data.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/loading-spinner Docs}
 */
export const LoadingSpinner = createAndRegisterRemoteReactComponent('LoadingSpinner');
/**
 * The `ProgressBar` component renders a visual indicator showing a numeric and/or percentage-based representation of progress. The percentage is calculated based on the maximum possible value specified in the component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/progress-bar Docs}
 */
export const ProgressBar = createAndRegisterRemoteReactComponent('ProgressBar');
/**
 * The `Select` component renders a dropdown menu select field where a user can select a single value. A search bar will be automatically included when there are more than seven options. Like other inputs, this component should be used within a `Form` that has a submit button.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/select Docs}
 */
export const Select = createAndRegisterRemoteReactComponent('Select');
/**
 * The `Tag` component renders a tag to label or categorize information or other components. Tags can be static or clickable for invoking functions.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/tag Docs}
 */
export const Tag = createAndRegisterRemoteReactComponent('Tag', {
    fragmentProps: ['overlay'],
});
/**
 * The `Text` component renders text with formatting options.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/text Docs}
 */
export const Text = createAndRegisterRemoteReactComponent('Text');
/**
 * The `Tile` component renders a square tile that can contain other components. Use this component to create groups of related components.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/tile Docs}
 */
export const Tile = createAndRegisterRemoteReactComponent('Tile');
/** @deprecated use Flex instead. It will be removed in the next release. */
export const Stack = createAndRegisterRemoteReactComponent('Stack');
/**
 * The `ToggleGroup` component renders a list of selectable options, either in radio button or checkbox form.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/toggle-group Docs}
 */
export const ToggleGroup = createAndRegisterRemoteReactComponent('ToggleGroup');
/**
 * The `StatisticsItem` component renders a single data point within a `Statistics` component. Use this component to display a single data point, such as a number or percentage.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/statistics Docs}
 */
export const StatisticsItem = createAndRegisterRemoteReactComponent('StatisticsItem');
/**
 * The `Statistics` component renders a visual spotlight of one or more data points. Includes the `StatisticsItem` and `StatisticsTrend` subcomponents.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/statistics Docs}
 */
export const Statistics = createAndRegisterRemoteReactComponent('Statistics');
/**
 * The `StatisticsTrend` component renders a percentage trend value and direction alonside a `StatisticsItem` component. Use this component within the `StatisticsItem` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/statistics Docs}
 */
export const StatisticsTrend = createAndRegisterRemoteReactComponent('StatisticsTrend');
/**
 * The `Table` component renders a table. To format the table, use the subcomponents `TableHead`, `TableRow`, `TableHeader`, `TableBody`, `TableCell`and `TableFooter`.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/table Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/design-patterns#table Design Pattern Example}
 */
export const Table = createAndRegisterRemoteReactComponent('Table');
/**
 * The `TableFooter` component renders a footer within a `Table` component. Use this component to display totals or other summary information.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/table Docs}
 */
export const TableFooter = createAndRegisterRemoteReactComponent('TableFooter');
/**
 * The `TableCell` component renders individual cells within the `TableBody` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/table Docs}
 */
export const TableCell = createAndRegisterRemoteReactComponent('TableCell');
/**
 * The `TableRow` component renders a row within the `TableBody` or `TableHead` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/table Docs}
 */
export const TableRow = createAndRegisterRemoteReactComponent('TableRow');
/**
 * The `TableBody` component renders the body (rows and cells) of a table within the `Table` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/table Docs}
 */
export const TableBody = createAndRegisterRemoteReactComponent('TableBody');
/**
 * The `TableHeader` component renders individual cells containing bolded column labels, within `TableHead`.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/table Docs}
 */
export const TableHeader = createAndRegisterRemoteReactComponent('TableHeader');
/**
 * The `TableHead` component renders the header section of the `Table` component, containing column labels.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/table Docs}
 */
export const TableHead = createAndRegisterRemoteReactComponent('TableHead');
/**
 * The `NumberInput` component renders a number input field. Like other inputs, this component should be used within a `Form` that has a submit button.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/number-input Docs}
 */
export const NumberInput = createAndRegisterRemoteReactComponent('NumberInput');
/**
 * The `Box` component renders an empty div container for fine tuning the spacing of components. Commonly used with the `Flex` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/box Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/flex-and-box Flex and Box Example}
 */
export const Box = createAndRegisterRemoteReactComponent('Box');
/**
 * The `StepIndicator` component renders an indicator to show the current step of a multi-step process.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/step-indicator Docs}
 */
export const StepIndicator = createAndRegisterRemoteReactComponent('StepIndicator');
/**
 * The `Accordion` component renders an expandable and collapsable section that can contain other components. This component can be helpful for saving space and breaking up extension content.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/accordion Docs}
 */
export const Accordion = createAndRegisterRemoteReactComponent('Accordion');
/**
 * The MultiSelect component renders a dropdown menu select field where a user can select multiple values. Commonly used within the `Form` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/multi-select Docs}
 */
export const MultiSelect = createAndRegisterRemoteReactComponent('MultiSelect');
/**
 * The `Flex` component renders a flex container that can contain other components, and arrange them with props. Use this component to create a flexible and responsive layout.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/flex Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/flex-and-box Flex and Box Example}
 */
export const Flex = createAndRegisterRemoteReactComponent('Flex');
/**
 * The `DateInput` component renders an input field where a user can select a date. Commonly used within the `Form` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/date-input Docs}
 */
export const DateInput = createAndRegisterRemoteReactComponent('DateInput');
/**
 * The `Checkbox` component renders a single checkbox input. Commonly used within the `Form` component. If you want to display multiple checkboxes, you should use `ToggleGroup` instead, as it comes with extra logic for handling multiple checkboxes.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/checkbox Docs}
 */
export const Checkbox = createAndRegisterRemoteReactComponent('Checkbox');
/**
 * The `RadioButton` component renders a single radio input. Commonly used within the `Form` component. If you want to display multiple radio inputs, you should use `ToggleGroup` instead, as it comes with extra logic for handling multiple inputs.
 */
export const RadioButton = createAndRegisterRemoteReactComponent('RadioButton');
/**
 * The `List` component renders a list of items. Use this component to display a list of items, such as a list of contacts, tasks, or other data. A list can be styled as a bulleted list or a numbered list.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/list Docs}
 */
export const List = createAndRegisterRemoteReactComponent('List');
/**
 * The `Toggle` component renders a boolean toggle switch that can be configured with sizing, label position, read-only, and more.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/toggle Docs}
 */
export const Toggle = createAndRegisterRemoteReactComponent('Toggle');
/**
 * The `Dropdown` component renders a dropdown menu that can appear as a button or hyperlink. Use this component to enable users to select from multiple options in a compact list.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/dropdown Docs}
 */
export const Dropdown = createAndRegisterRemoteCompoundReactComponent('Dropdown', {
    compoundComponentProperties: {
        /**
         * The `Dropdown.ButtonItem` component represents a single option within a `Dropdown` menu. Use this component as a child of the `Dropdown` component.
         *
         * **Links:**
         *
         * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/dropdown Docs}
         */
        ButtonItem: createAndRegisterRemoteReactComponent('DropdownButtonItem', {
            fragmentProps: ['overlay'],
        }),
    },
});
/**
 * The Panel component renders a panel overlay on the right side of the page and contains other components.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/panel Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/overlay-example Overlay Example}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/design-patterns#panel Design Pattern Examples}
 */
export const Panel = createAndRegisterRemoteReactComponent('Panel');
/**
 * The `PanelFooter` is a sticky footer component displayed at the bottom of a `Panel` component. Use this component to display actions or other content that should be visible at all times. Include only one `PanelFooter` component per `Panel`.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/panel-footer Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/overlay-example Overlay Example}
 */
export const PanelFooter = createAndRegisterRemoteReactComponent('PanelFooter');
/**
 * The `PanelBody` component is a container that wraps the panel's content and makes it scrollable. Include only one `PanelBody` component per `Panel`.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/panel-footer Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/overlay-example Overlay Example}
 */
export const PanelBody = createAndRegisterRemoteReactComponent('PanelBody');
/**
 * The `PanelSection` component is a container that adds padding and bottom margin to provide spacing between content. Use the `PanelSection` component to separate content within a `PanelBody`.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/panel-footer Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/overlay-example Overlay Example}
 */
export const PanelSection = createAndRegisterRemoteReactComponent('PanelSection');
/**
 * The `StepperInput` component renders a number input field that can be increased or decreased by a set number. Commonly used within the `Form` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/stepper-input Docs}
 */
export const StepperInput = createAndRegisterRemoteReactComponent('StepperInput');
/**
 * The Modal component renders a pop-up overlay that can contain other components.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/modal Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/overlay-example Overlay Example}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/design-patterns#modal Design Pattern Examples}
 */
export const Modal = createAndRegisterRemoteReactComponent('Modal');
/**
 * The `ModalBody` component contains the main content of the modal. One `ModalBody` is required per `Modal`.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/modal Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/overlay-example Overlay Example}
 */
export const ModalBody = createAndRegisterRemoteReactComponent('ModalBody');
/**
 * The `ModalFooter` component is an optional component to format the footer section of the modal. Use one `ModalFooter` per `Modal`.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/modal Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/overlay-example Overlay Example}
 */
export const ModalFooter = createAndRegisterRemoteReactComponent('ModalFooter');
/**
 * Use the `Icon` component to render a visual icon within other components. It can generally be used inside most components, excluding ones that don't support child components.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/icon Docs}
 */
export const Icon = createAndRegisterRemoteReactComponent('Icon');
/**
 * The `StatusTag` component renders a visual indicator to display the current status of an item. Status tags can be static or clickable.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/status-tag Docs}
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/status-tag#variants Variants}
 */
export const StatusTag = createAndRegisterRemoteReactComponent('StatusTag');
/**
 * The `LoadingButton` component renders a button with loading state options.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/loading-button Docs}
 */
export const LoadingButton = createAndRegisterRemoteReactComponent('LoadingButton', {
    fragmentProps: ['overlay'],
});
/**
 * The `BarChart` component renders a bar chart for visualizing data. This type of chart is best suited for comparing categorical data.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/bar-chart BarChart Docs}
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/charts Charts Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/charts-example Charts Example}
 */
export const BarChart = createAndRegisterRemoteReactComponent('BarChart');
/**
 * The `LineChart` component renders a line chart for visualizing data. This type of chart is best suited for time series plots or trend data.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/line-chart LineChart Docs}
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/charts Charts Docs}
 * - {@link https://github.com/HubSpot/ui-extensions-examples/tree/main/charts-example Charts Example}
 */
export const LineChart = createAndRegisterRemoteReactComponent('LineChart');
/**
 * The `ScoreCircle` component displays a score value (0-100) as a circular progress indicator with color-coded bands.
 * Scores are color-coded: 0-32 (alert/red), 33-65 (warning/yellow), 66-100 (success/green).
 * @example
 * ```tsx
 *   <ScoreCircle score={75} />
 * ```
 */
export const ScoreCircle = createAndRegisterRemoteReactComponent('ScoreCircle');
/**
 * `Tabs` allow you to group related content in a compact space, allowing users to switch between views without leaving the page.
 * @example
 * ```tsx
 * <Tabs defaultSelected="1">
 *   <Tab tabId="1">First tab content</Tab>
 *   <Tab tabId="2">Second tab content</Tab>
 * </Tabs>
 * ```
 *
 * **Links:**
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/tabs Documentation}
 * - {@link https://github.com/hubspotdev/uie-tabbed-product-carousel Tabs Example}
 */
export const Tabs = createAndRegisterRemoteReactComponent('Tabs');
/**
 * Each `Tab` represents a single tab (or "view") within the parent `Tabs` component.
 * @example
 * ```tsx
 * <Tabs defaultSelected="1">
 *   <Tab tabId="1">First tab content</Tab>
 *   <Tab tabId="2">Second tab content</Tab>
 * </Tabs>
 * ```
 *
 * **Links:**
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/tabs Documentation}
 * - {@link https://github.com/hubspotdev/uie-tabbed-product-carousel Tabs Example}
 */
export const Tab = createAndRegisterRemoteReactComponent('Tab');
/**
 * The `Illustration` component renders an illustration.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/illustration Illustration Docs}
 */
export const Illustration = createAndRegisterRemoteReactComponent('Illustration');
/**
 * The `Tooltip` component renders a tooltip for a component.
 *
 * **Links:**
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/tooltip Documentation}
 */
export const Tooltip = createAndRegisterRemoteReactComponent('Tooltip');
/**
 * The `SearchInput` component renders a search input field.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/search-input SearchInput Docs}
 */
export const SearchInput = createAndRegisterRemoteReactComponent('SearchInput');
/**
 * The `TimeInput` component renders an input field where a user can select a time. Commonly used within the `Form` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/time-input Docs}
 */
export const TimeInput = createAndRegisterRemoteReactComponent('TimeInput');
/**
 * The `CurrencyInput` component renders a currency input field with proper formatting,
 * currency symbols, and locale-specific display patterns. Commonly used within the `Form` component.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/currency-input Docs}
 */
export const CurrencyInput = createAndRegisterRemoteReactComponent('CurrencyInput');
/**
 * The `Inline` component spreads aligns its children horizontally (along the x-axis).
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/inline Docs}
 */ export const Inline = createAndRegisterRemoteReactComponent('Inline');
/**
 * The `AutoGrid` component renders a responsive grid layout that automatically adjusts the number of columns based on available space. Use this component to create flexible grid layouts for cards, tiles, or other content.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/simple-grid Docs}
 */
export const AutoGrid = createAndRegisterRemoteReactComponent('AutoGrid');
////////////////////////////////////////////////////////
// CRM COMPONENTS
////////////////////////////////////////////////////////
export const CrmPropertyList = createAndRegisterRemoteReactComponent('CrmPropertyList');
export const CrmAssociationTable = createAndRegisterRemoteReactComponent('CrmAssociationTable');
export const CrmDataHighlight = createAndRegisterRemoteReactComponent('CrmDataHighlight');
export const CrmReport = createAndRegisterRemoteReactComponent('CrmReport');
export const CrmAssociationPivot = createAndRegisterRemoteReactComponent('CrmAssociationPivot');
export const CrmAssociationPropertyList = createAndRegisterRemoteReactComponent('CrmAssociationPropertyList');
export const CrmAssociationStageTracker = createAndRegisterRemoteReactComponent('CrmAssociationStageTracker');
export const CrmSimpleDeadline = createAndRegisterRemoteReactComponent('CrmSimpleDeadline');
export const CrmStageTracker = createAndRegisterRemoteReactComponent('CrmStageTracker');
export const CrmStatistics = createAndRegisterRemoteReactComponent('CrmStatistics');
export const CrmActionButton = createAndRegisterRemoteReactComponent('CrmActionButton');
export const CrmActionLink = createAndRegisterRemoteReactComponent('CrmActionLink');
export const CrmCardActions = createAndRegisterRemoteReactComponent('CrmCardActions');
////////////////////////////////////////////////////////
// APP HOME COMPONENTS
////////////////////////////////////////////////////////
/**
 * The `HeaderActions` component renders a container for action buttons in the app home header. It accepts `PrimaryHeaderActionButton` and `SecondaryHeaderActionButton` as children.
 *
 */
export const HeaderActions = createAndRegisterRemoteReactComponent('HeaderActions');
/**
 * The `PrimaryHeaderActionButton` component renders a primary action button in the app home header. This button is styled as the main call-to-action and only one should be used per `HeaderActions` container.
 *
 */
export const PrimaryHeaderActionButton = createAndRegisterRemoteReactComponent('PrimaryHeaderActionButton', {
    fragmentProps: ['overlay'],
});
/**
 * The `SecondaryHeaderActionButton` component renders a secondary action button in the app home header. Multiple secondary actions can be used and they will be grouped appropriately in the header.
 *
 */
export const SecondaryHeaderActionButton = createAndRegisterRemoteReactComponent('SecondaryHeaderActionButton', {
    fragmentProps: ['overlay'],
});
////////////////////////////////////////////////////////
// APP PAGE COMPONENTS
////////////////////////////////////////////////////////
/**
 * The `PageHeader` component renders the actions within the header of the page. It accepts `PrimaryAction` and `SecondaryActions` as children.
 *
 * **Links:**
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/app-page-components/page-header Docs}
 */
export const PageHeader = createAndRegisterRemoteCompoundReactComponent('PageHeader', {
    compoundComponentProperties: {
        PrimaryAction: createAndRegisterRemoteReactComponent('PageHeaderPrimaryAction'),
        SecondaryActions: createAndRegisterRemoteReactComponent('PageHeaderSecondaryActions'),
        Link: createAndRegisterRemoteReactComponent('PageHeaderLink'),
        PageLink: createAndRegisterRemoteReactComponent('PageHeaderPageLink'),
    },
});
/**
 * The 'PageBreadcrumbs' component renders a list of links to show the user's current location within the app and allow them to navigate back to previous pages.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/app-page-components/page-breadcrumbs Docs}
 */
export const PageBreadcrumbs = createAndRegisterRemoteCompoundReactComponent('PageBreadcrumbs', {
    compoundComponentProperties: {
        PageLink: createAndRegisterRemoteReactComponent('PageBreadcrumbsPageLink'),
        Current: createAndRegisterRemoteReactComponent('PageBreadcrumbsCurrent'),
    },
});
export const PageLink = createAndRegisterRemoteReactComponent('PageLink');
export const PageTitle = createAndRegisterRemoteReactComponent('PageTitle');
////////////////////////////////////////////////////////
// EXPERIMENTAL COMPONENTS
////////////////////////////////////////////////////////
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const Iframe = createAndRegisterRemoteReactComponent('Iframe');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const MediaObject = createAndRegisterRemoteReactComponent('MediaObject', {
    fragmentProps: ['itemRight', 'itemLeft'],
});
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const Stack2 = createAndRegisterRemoteReactComponent('Stack2');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const Center = createAndRegisterRemoteReactComponent('Center');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const Grid = createAndRegisterRemoteReactComponent('Grid');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const GridItem = createAndRegisterRemoteReactComponent('GridItem');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const SettingsView = createAndRegisterRemoteReactComponent('SettingsView');
/**
 * The `ExpandableText` component renders a text that can be expanded or collapsed based on a maximum height.
 *
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/expandable-text ExpandableText Docs}
 */
export const ExpandableText = createAndRegisterRemoteReactComponent('ExpandableText');
/**
 * The `Popover` component renders a popover overlay that can contain other components.
 *
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 *
 * **Links:**
 *
 * - {@link https://developers.hubspot.com/docs/reference/ui-components/standard-components/popover Popover Docs}
 */
export const Popover = createAndRegisterRemoteReactComponent('Popover');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const FileInput = createAndRegisterRemoteReactComponent('FileInput');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const FileUpload = createAndRegisterRemoteReactComponent('FileUpload');
/**
 * @experimental This component is experimental. Avoid using it in production due to potential breaking changes. Your feedback is valuable for improvements. Stay tuned for updates.
 */
export const FileViewer = createAndRegisterRemoteReactComponent('FileViewer');
////////////////////////////////////////////////////////
// EXPERIMENTAL SPACING VARIANTS
////////////////////////////////////////////////////////
export const ExperimentalFlex = createAndRegisterRemoteReactComponent('ExperimentalFlex');
export const ExperimentalBox = createAndRegisterRemoteReactComponent('ExperimentalBox');
export const ExperimentalInline = createAndRegisterRemoteReactComponent('ExperimentalInline');
export const ExperimentalAutoGrid = createAndRegisterRemoteReactComponent('ExperimentalAutoGrid');
export const ExperimentalButtonRow = createAndRegisterRemoteReactComponent('ExperimentalButtonRow');
export const ExperimentalButton = createAndRegisterRemoteReactComponent('ExperimentalButton', {
    fragmentProps: ['overlay'],
});
export const ExperimentalLoadingButton = createAndRegisterRemoteReactComponent('ExperimentalLoadingButton', {
    fragmentProps: ['overlay'],
});
export const ExperimentalInput = createAndRegisterRemoteReactComponent('ExperimentalInput');
export const ExperimentalTextArea = createAndRegisterRemoteReactComponent('ExperimentalTextArea');
export const ExperimentalNumberInput = createAndRegisterRemoteReactComponent('ExperimentalNumberInput');
export const ExperimentalDateInput = createAndRegisterRemoteReactComponent('ExperimentalDateInput');
export const ExperimentalTimeInput = createAndRegisterRemoteReactComponent('ExperimentalTimeInput');
export const ExperimentalCurrencyInput = createAndRegisterRemoteReactComponent('ExperimentalCurrencyInput');
export const ExperimentalStepperInput = createAndRegisterRemoteReactComponent('ExperimentalStepperInput');
export const ExperimentalSearchInput = createAndRegisterRemoteReactComponent('ExperimentalSearchInput');
export const ExperimentalSelect = createAndRegisterRemoteReactComponent('ExperimentalSelect');
export const ExperimentalMultiSelect = createAndRegisterRemoteReactComponent('ExperimentalMultiSelect');
export const ExperimentalCheckbox = createAndRegisterRemoteReactComponent('ExperimentalCheckbox');
export const ExperimentalRadioButton = createAndRegisterRemoteReactComponent('ExperimentalRadioButton');
export const ExperimentalToggle = createAndRegisterRemoteReactComponent('ExperimentalToggle');
