// @hubspot/api-client/lib/codegen/crm/exports/types/ObservableAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObservablePublicExportsApi = void 0;
const rxjsStub_1 = require("../rxjsStub");
const rxjsStub_2 = require("../rxjsStub");
const PublicExportsApi_1 = require("../apis/PublicExportsApi");
class ObservablePublicExportsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.configuration = configuration;
        this.requestFactory = requestFactory || new PublicExportsApi_1.PublicExportsApiRequestFactory(configuration);
        this.responseProcessor = responseProcessor || new PublicExportsApi_1.PublicExportsApiResponseProcessor();
    }
    getStatusWithHttpInfo(taskId, _options) {
        const requestContextPromise = this.requestFactory.getStatus(taskId, _options);
        let middlewarePreObservable = (0, rxjsStub_1.from)(requestContextPromise);
        for (let middleware of this.configuration.middleware) {
            middlewarePreObservable = middlewarePreObservable.pipe((0, rxjsStub_2.mergeMap)((ctx) => middleware.pre(ctx)));
        }
        return middlewarePreObservable.pipe((0, rxjsStub_2.mergeMap)((ctx) => this.configuration.httpApi.send(ctx))).
            pipe((0, rxjsStub_2.mergeMap)((response) => {
            let middlewarePostObservable = (0, rxjsStub_1.of)(response);
            for (let middleware of this.configuration.middleware) {
                middlewarePostObservable = middlewarePostObservable.pipe((0, rxjsStub_2.mergeMap)((rsp) => middleware.post(rsp)));
            }
            return middlewarePostObservable.pipe((0, rxjsStub_2.map)((rsp) => this.responseProcessor.getStatusWithHttpInfo(rsp)));
        }));
    }
    getStatus(taskId, _options) {
        return this.getStatusWithHttpInfo(taskId, _options).pipe((0, rxjsStub_2.map)((apiResponse) => apiResponse.data));
    }
    startWithHttpInfo(publicExportRequest, _options) {
        const requestContextPromise = this.requestFactory.start(publicExportRequest, _options);
        let middlewarePreObservable = (0, rxjsStub_1.from)(requestContextPromise);
        for (let middleware of this.configuration.middleware) {
            middlewarePreObservable = middlewarePreObservable.pipe((0, rxjsStub_2.mergeMap)((ctx) => middleware.pre(ctx)));
        }
        return middlewarePreObservable.pipe((0, rxjsStub_2.mergeMap)((ctx) => this.configuration.httpApi.send(ctx))).
            pipe((0, rxjsStub_2.mergeMap)((response) => {
            let middlewarePostObservable = (0, rxjsStub_1.of)(response);
            for (let middleware of this.configuration.middleware) {
                middlewarePostObservable = middlewarePostObservable.pipe((0, rxjsStub_2.mergeMap)((rsp) => middleware.post(rsp)));
            }
            return middlewarePostObservable.pipe((0, rxjsStub_2.map)((rsp) => this.responseProcessor.startWithHttpInfo(rsp)));
        }));
    }
    start(publicExportRequest, _options) {
        return this.startWithHttpInfo(publicExportRequest, _options).pipe((0, rxjsStub_2.map)((apiResponse) => apiResponse.data));
    }
}
exports.ObservablePublicExportsApi = ObservablePublicExportsApi;
