// @hubspot/api-client/lib/codegen/crm/exports/types/PromiseAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromisePublicExportsApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class PromisePublicExportsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservablePublicExportsApi(configuration, requestFactory, responseProcessor);
    }
    getStatusWithHttpInfo(taskId, _options) {
        const result = this.api.getStatusWithHttpInfo(taskId, _options);
        return result.toPromise();
    }
    getStatus(taskId, _options) {
        const result = this.api.getStatus(taskId, _options);
        return result.toPromise();
    }
    startWithHttpInfo(publicExportRequest, _options) {
        const result = this.api.startWithHttpInfo(publicExportRequest, _options);
        return result.toPromise();
    }
    start(publicExportRequest, _options) {
        const result = this.api.start(publicExportRequest, _options);
        return result.toPromise();
    }
}
exports.PromisePublicExportsApi = PromisePublicExportsApi;
