// @hubspot/ui-extensions/dist/testing/internal/mocks/index.js
import { act as reactAct, useSyncExternalStore } from 'react';
import { spyOn } from 'tinyspy';
import { createMocksAppPageLocationStore, } from "./mock-app-page-location.js";
import { createMockExtensionPointApi } from "./mock-extension-point-api.js";
import { createMockHooks } from "./mock-hooks.js";
/**
 * This function replaces all methods with a spy using the `spyOn` function from the `tinyspy` library.
 * The object tree is walked recursively to replace all methods with spies. The original object is not modified.
 *
 * @param object The object to spy on.
 * @returns A new object with all of the methods replaced with spies.
 */
const spyOnAllFunctionsRecursive = (object) => {
    if (!object || typeof object !== 'object' || Array.isArray(object)) {
        return object;
    }
    const objectWithSpies = Object.create(null);
    for (const [key, value] of Object.entries(object)) {
        if (typeof value === 'function') {
            objectWithSpies[key] = spyOn(object, key);
        }
        else {
            objectWithSpies[key] = spyOnAllFunctionsRecursive(value);
        }
    }
    return objectWithSpies;
};
/**
 * Creates a set of mocks and spies for the various functions such as the `useCrmProperties` and `useAssociations`
 * hook functions.
 *
 * @returns An object containing the mocks and spies.
 */
export const createRendererMocksContext = (extensionPointLocation) => {
    const mockExtensionPointApi = createMockExtensionPointApi(extensionPointLocation);
    // Build a mocks object with the default mock functions.
    // NOTE: The mock implementations can be overridden by tests using the spies that are provided in the returned `spies` object.
    const mocks = {
        ...createMockHooks(),
        ...mockExtensionPointApi,
    };
    // Recursively create spies for all functions in the mocks object.
    const mocksWithSpies = spyOnAllFunctionsRecursive(mocks);
    const appPageLocationStore = createMocksAppPageLocationStore();
    return {
        mocks: {
            // NOTE: These mocks are exposed as part of our public API.
            ...mocksWithSpies,
            setPageLocation: (location) => {
                reactAct(() => {
                    appPageLocationStore.setPageLocation(location);
                });
            },
        },
        appPageLocationStore,
        useAppPageLocation: () => {
            return useSyncExternalStore(appPageLocationStore.subscribe, appPageLocationStore.getSnapshot);
        },
    };
};
