// @hubspot/ui-extensions/dist/testing/internal/mocks/mock-hooks.js
import { useMocksContext } from "../../../internal/hook-utils.js";
/**
 * Creates a record of property names and their fake values.
 *
 * @param propertyNames The names of the properties to create to create fake properties.
 * @returns A record of property names and their fake values.
 */
const createFakeProperties = (propertyNames) => {
    if (!propertyNames || propertyNames.length === 0) {
        return {};
    }
    return propertyNames.reduce((acc, propertyName) => {
        acc[propertyName] = `fake_${propertyName}`;
        return acc;
    }, {});
};
export const createMockHooks = () => {
    const useRequiredMocksContext = () => {
        const mocksContext = useMocksContext();
        if (!mocksContext) {
            throw new Error('Illegal State: Mocks context not found.');
        }
        return mocksContext;
    };
    // Build a mocks object with the default mock functions.
    // NOTE: The mock implementations can be overridden by tests using the spies that are provided in the returned `spies` object.
    return {
        useCrmProperties: (propertyNames) => {
            const properties = createFakeProperties(propertyNames);
            return {
                properties,
                error: null,
                isLoading: false,
                isRefetching: false,
                refetch: async () => { },
            };
        },
        useAssociations: (config) => {
            const { properties: propertyNames } = config;
            const properties = createFakeProperties(propertyNames);
            return {
                results: [
                    {
                        toObjectId: 123,
                        associationTypes: [
                            {
                                category: 'HUBSPOT_DEFINED',
                                typeId: 1,
                                label: 'Primary',
                            },
                        ],
                        properties,
                    },
                ],
                error: null,
                isLoading: false,
                isRefetching: false,
                refetch: async () => { },
                pagination: {
                    hasNextPage: false,
                    hasPreviousPage: false,
                    currentPage: 1,
                    pageSize: 10,
                    nextPage: () => { },
                    previousPage: () => { },
                    reset: () => { },
                },
            };
        },
        useCrmSearch: (config) => {
            const { properties: propertyNames } = config;
            const properties = createFakeProperties(propertyNames);
            return {
                results: [
                    {
                        objectId: 456,
                        properties,
                    },
                ],
                total: 1,
                error: null,
                isLoading: false,
                isRefetching: false,
                refetch: async () => { },
                pagination: {
                    hasNextPage: false,
                    hasPreviousPage: false,
                    currentPage: 1,
                    pageSize: 10,
                    nextPage: () => { },
                    previousPage: () => { },
                    reset: () => { },
                },
            };
        },
        useExtensionContext: () => {
            const { mocks } = useRequiredMocksContext();
            return mocks.context;
        },
        useExtensionActions: () => {
            const { mocks } = useRequiredMocksContext();
            return mocks.actions;
        },
        useExtensionApi: () => {
            const { mocks } = useRequiredMocksContext();
            return {
                context: mocks.context,
                actions: mocks.actions,
                runServerlessFunction: mocks.runServerlessFunction,
            };
        },
    };
};
