// @hubspot/ui-extensions/dist/testing/internal/mocks/mock-extension-point-api.js
import { ServerlessExecutionStatus } from "../../../shared/types/http-requests.js";
const fakePrefix = 'fake_';
const createMockServerlessFuncRunner = () => {
    return () => Promise.resolve({
        status: ServerlessExecutionStatus.Success,
        response: {
            fakeData: 'hello',
        },
    });
};
const createMockCrmHostActions = () => {
    return {
        addAlert: () => { },
        reloadPage: () => { },
        fetchCrmObjectProperties: () => Promise.resolve({}),
        openIframeModal: () => { },
        refreshObjectProperties: () => { },
        onCrmPropertiesUpdate: () => { },
    };
};
const createMockUiePlatformActions = () => {
    return {
        copyTextToClipboard: async () => { },
        closeOverlay: () => { },
        navigateToPage: () => { },
        openIframeModal: () => { },
        reloadPage: () => { },
    };
};
const createMockSettingsActions = () => {
    return {
        addAlert: () => { },
    };
};
const createMockAppHomeActions = () => {
    return {
        addAlert: () => { },
    };
};
const createMockPagesActions = () => {
    return {
        addAlert: () => { },
    };
};
const createFakeEmail = () => `${fakePrefix}email@example.com`;
const createFakeFirstName = () => `${fakePrefix}firstName`;
const createFakeLastName = () => `${fakePrefix}lastName`;
const createFakeUser = () => {
    return {
        id: 1,
        emails: [createFakeEmail()],
        email: createFakeEmail(),
        firstName: createFakeFirstName(),
        lastName: createFakeLastName(),
        roles: [],
        teams: [],
        permissions: [],
    };
};
const createFakePortalContext = () => {
    return {
        id: 123,
        timezone: 'America/New_York',
    };
};
const createFakeCrmRecordContext = () => {
    return {
        objectId: 123,
        objectTypeId: '0-1',
    };
};
const createFakeCrmContext = (extensionPointLocationName) => {
    return {
        location: extensionPointLocationName,
        user: createFakeUser(),
        portal: createFakePortalContext(),
        crm: createFakeCrmRecordContext(),
        variables: {
            fake_variable: 'fake_value',
        },
    };
};
const mockCreators = {
    'crm.record.tab': () => {
        return {
            runServerlessFunction: createMockServerlessFuncRunner(),
            actions: {
                ...createMockCrmHostActions(),
                ...createMockUiePlatformActions(),
            },
            context: createFakeCrmContext('crm.record.tab'),
        };
    },
    'crm.preview': () => {
        return {
            runServerlessFunction: createMockServerlessFuncRunner(),
            actions: {
                ...createMockCrmHostActions(),
                ...createMockUiePlatformActions(),
            },
            context: createFakeCrmContext('crm.preview'),
        };
    },
    'crm.record.sidebar': () => {
        return {
            runServerlessFunction: createMockServerlessFuncRunner(),
            actions: {
                ...createMockCrmHostActions(),
                ...createMockUiePlatformActions(),
            },
            context: createFakeCrmContext('crm.record.sidebar'),
        };
    },
    'helpdesk.sidebar': () => {
        return {
            runServerlessFunction: createMockServerlessFuncRunner(),
            actions: {
                ...createMockCrmHostActions(),
                ...createMockUiePlatformActions(),
            },
            context: createFakeCrmContext('helpdesk.sidebar'),
        };
    },
    'uie.playground.middle': () => {
        return {
            runServerlessFunction: createMockServerlessFuncRunner(),
            actions: {
                warn: () => { },
            },
            context: {
                location: 'uie.playground.middle',
                user: createFakeUser(),
                portal: createFakePortalContext(),
            },
        };
    },
    settings: () => {
        return {
            runServerlessFunction: createMockServerlessFuncRunner(),
            actions: {
                ...createMockSettingsActions(),
                ...createMockUiePlatformActions(),
            },
            context: {
                location: 'settings',
                user: createFakeUser(),
                portal: createFakePortalContext(),
            },
        };
    },
    home: () => {
        return {
            runServerlessFunction: createMockServerlessFuncRunner(),
            actions: {
                ...createMockAppHomeActions(),
                ...createMockUiePlatformActions(),
            },
            context: {
                location: 'home',
                user: createFakeUser(),
                portal: createFakePortalContext(),
            },
        };
    },
    pages: () => {
        return {
            runServerlessFunction: createMockServerlessFuncRunner(),
            actions: {
                ...createMockPagesActions(),
                ...createMockUiePlatformActions(),
            },
            context: {
                location: 'pages',
                user: createFakeUser(),
                portal: createFakePortalContext(),
            },
        };
    },
};
export const createMockExtensionPointApi = (extensionPointLocation) => {
    const mockCreator = mockCreators[extensionPointLocation];
    if (!mockCreator) {
        throw new Error(`No mock creator found for extension point location: ${extensionPointLocation}`);
    }
    return mockCreator();
};
