// @hubspot/api-client/lib/codegen/crm/exports/models/PublicExportRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublicExportRequestExportInternalValuesOptionsEnum = exports.PublicExportRequestLanguageEnum = exports.PublicExportRequestFormatEnum = exports.PublicExportRequestExportTypeEnum = exports.PublicExportRequest = void 0;
class PublicExportRequest {
    static getAttributeTypeMap() {
        return PublicExportRequest.attributeTypeMap;
    }
    constructor() {
    }
}
exports.PublicExportRequest = PublicExportRequest;
PublicExportRequest.discriminator = undefined;
PublicExportRequest.attributeTypeMap = [
    {
        "name": "exportType",
        "baseName": "exportType",
        "type": "PublicExportRequestExportTypeEnum",
        "format": ""
    },
    {
        "name": "format",
        "baseName": "format",
        "type": "PublicExportRequestFormatEnum",
        "format": ""
    },
    {
        "name": "exportName",
        "baseName": "exportName",
        "type": "string",
        "format": ""
    },
    {
        "name": "objectProperties",
        "baseName": "objectProperties",
        "type": "Array<string>",
        "format": ""
    },
    {
        "name": "associatedObjectType",
        "baseName": "associatedObjectType",
        "type": "string",
        "format": ""
    },
    {
        "name": "objectType",
        "baseName": "objectType",
        "type": "string",
        "format": ""
    },
    {
        "name": "language",
        "baseName": "language",
        "type": "PublicExportRequestLanguageEnum",
        "format": ""
    },
    {
        "name": "exportInternalValuesOptions",
        "baseName": "exportInternalValuesOptions",
        "type": "Array<PublicExportRequestExportInternalValuesOptionsEnum>",
        "format": ""
    },
    {
        "name": "overrideAssociatedObjectsPerDefinitionPerRowLimit",
        "baseName": "overrideAssociatedObjectsPerDefinitionPerRowLimit",
        "type": "boolean",
        "format": ""
    },
    {
        "name": "publicCrmSearchRequest",
        "baseName": "publicCrmSearchRequest",
        "type": "PublicCrmSearchRequest",
        "format": ""
    },
    {
        "name": "listId",
        "baseName": "listId",
        "type": "string",
        "format": ""
    }
];
var PublicExportRequestExportTypeEnum;
(function (PublicExportRequestExportTypeEnum) {
    PublicExportRequestExportTypeEnum["List"] = "LIST";
})(PublicExportRequestExportTypeEnum = exports.PublicExportRequestExportTypeEnum || (exports.PublicExportRequestExportTypeEnum = {}));
var PublicExportRequestFormatEnum;
(function (PublicExportRequestFormatEnum) {
    PublicExportRequestFormatEnum["Xls"] = "XLS";
    PublicExportRequestFormatEnum["Xlsx"] = "XLSX";
    PublicExportRequestFormatEnum["Csv"] = "CSV";
})(PublicExportRequestFormatEnum = exports.PublicExportRequestFormatEnum || (exports.PublicExportRequestFormatEnum = {}));
var PublicExportRequestLanguageEnum;
(function (PublicExportRequestLanguageEnum) {
    PublicExportRequestLanguageEnum["En"] = "EN";
    PublicExportRequestLanguageEnum["De"] = "DE";
    PublicExportRequestLanguageEnum["Es"] = "ES";
    PublicExportRequestLanguageEnum["Fr"] = "FR";
    PublicExportRequestLanguageEnum["Ja"] = "JA";
    PublicExportRequestLanguageEnum["Nl"] = "NL";
    PublicExportRequestLanguageEnum["PtBr"] = "PT_BR";
    PublicExportRequestLanguageEnum["It"] = "IT";
    PublicExportRequestLanguageEnum["Pl"] = "PL";
    PublicExportRequestLanguageEnum["Sv"] = "SV";
    PublicExportRequestLanguageEnum["Fi"] = "FI";
    PublicExportRequestLanguageEnum["ZhTw"] = "ZH_TW";
    PublicExportRequestLanguageEnum["DaDk"] = "DA_DK";
    PublicExportRequestLanguageEnum["No"] = "NO";
})(PublicExportRequestLanguageEnum = exports.PublicExportRequestLanguageEnum || (exports.PublicExportRequestLanguageEnum = {}));
var PublicExportRequestExportInternalValuesOptionsEnum;
(function (PublicExportRequestExportInternalValuesOptionsEnum) {
    PublicExportRequestExportInternalValuesOptionsEnum["Names"] = "NAMES";
    PublicExportRequestExportInternalValuesOptionsEnum["Values"] = "VALUES";
})(PublicExportRequestExportInternalValuesOptionsEnum = exports.PublicExportRequestExportInternalValuesOptionsEnum || (exports.PublicExportRequestExportInternalValuesOptionsEnum = {}));
