// @hubspot/eslint-config-ui-extensions/dist/index.js
import globals from 'globals';
import { noBrowserDialogs } from "./rules/no-browser-dialogs.js";
import { noBrowserStorage } from "./rules/no-browser-storage.js";
import { noConsole } from "./rules/no-console.js";
import { noDomAccess } from "./rules/no-dom-access.js";
import { noExperimentalImports } from "./rules/no-experimental-imports.js";
import { unsupportedHtmlElements } from "./rules/no-html-elements.js";
import { noInvalidExtensionPointImports } from "./rules/no-invalid-extension-point-imports.js";
import { noInvalidImageSrc } from "./rules/no-invalid-image-src.js";
import { noNativeHttp } from "./rules/no-native-http.js";
import { noParentImports } from "./rules/no-parent-imports.js";
import { noRestrictedGlobals } from "./rules/no-restricted-globals.js";
import { languageOptions } from "./runtime.js";
const rules = {
    'no-html-elements': unsupportedHtmlElements,
    'no-invalid-extension-point-imports': noInvalidExtensionPointImports,
    'no-invalid-image-src': noInvalidImageSrc,
    'no-native-http': noNativeHttp,
    'no-browser-dialogs': noBrowserDialogs,
    'no-console': noConsole,
    'no-dom-access': noDomAccess,
    'no-experimental-imports': noExperimentalImports,
    'no-browser-storage': noBrowserStorage,
    'no-parent-imports': noParentImports,
    'no-restricted-globals': noRestrictedGlobals,
};
// We export a config array for our users to extend
export const config = [
    // Config for ESLint config files
    {
        files: ['.eslintrc.{js,cjs}', 'eslint.config.js'],
        languageOptions: {
            globals: globals.node,
            sourceType: 'script',
        },
    },
    // Base config for JS/TS files
    {
        files: ['**/*.{js,jsx,ts,tsx}'],
        languageOptions: {
            ...languageOptions,
            parserOptions: {
                ecmaFeatures: {
                    jsx: true,
                },
            },
        },
        plugins: {
            '@hubspot/ui-extensions': {
                rules,
            },
        },
        rules: {
            '@hubspot/ui-extensions/no-html-elements': 'error',
            '@hubspot/ui-extensions/no-invalid-extension-point-imports': 'error',
            '@hubspot/ui-extensions/no-invalid-image-src': 'error',
            '@hubspot/ui-extensions/no-native-http': 'error',
            '@hubspot/ui-extensions/no-browser-dialogs': 'error',
            '@hubspot/ui-extensions/no-console': 'warn',
            '@hubspot/ui-extensions/no-experimental-imports': 'warn',
            '@hubspot/ui-extensions/no-dom-access': 'error',
            '@hubspot/ui-extensions/no-browser-storage': 'error',
            '@hubspot/ui-extensions/no-parent-imports': 'error',
            '@hubspot/ui-extensions/no-restricted-globals': 'error',
        },
    },
];
// We also export a plugin object for tooling to use
const plugin = {
    rules,
    configs: {
        default: config,
    },
};
export default plugin;
