// @hubspot/ui-extensions/dist/experimental/api-client/errors.js
import { VALID_HTTP_METHODS } from "../../shared/types/api-client.js";
const DEFAULT_FORMAT_MAX_LENGTH = 280;
export const formatInvalidInput = (value, maxLength = DEFAULT_FORMAT_MAX_LENGTH) => {
    let s;
    try {
        s = String(JSON.stringify(value));
    }
    catch {
        s = '[object could not be serialized to JSON]';
    }
    if (s.length > maxLength) {
        return `${s.slice(0, maxLength - 1)}…`;
    }
    return s;
};
export class InvalidApiPathError extends Error {
    constructor(received) {
        super(`path must be a non-empty string that starts with /; received: ${formatInvalidInput(received)}`);
        this.name = 'InvalidApiPathError';
    }
}
export class InvalidApiMethodError extends Error {
    constructor(method) {
        super(`${String(method)} is not valid. method must be one of ${[...VALID_HTTP_METHODS].join(', ')}`);
        this.name = 'InvalidApiMethodError';
    }
}
export class InvalidApiHeadersError extends Error {
    constructor(receivedHeaders) {
        super(`headers must be an object whose values are all strings; received: ${formatInvalidInput(receivedHeaders)}`);
        this.name = 'InvalidApiHeadersError';
    }
}
export class InvalidApiContentTypeError extends Error {
    constructor(received) {
        super(`contentType must be a string; received: ${formatInvalidInput(received)}`);
        this.name = 'InvalidApiContentTypeError';
    }
}
export class InvalidApiBodyError extends Error {
    constructor(received) {
        super(`body must be a plain object; received: ${formatInvalidInput(received)}`);
        this.name = 'InvalidApiBodyError';
    }
}
