// @hubspot/ui-extensions/dist/shared/types/shared.js
export class ExtensionEvent {
    type;
    bubbles;
    timeStamp;
    id;
    constructor(event) {
        this.bubbles = event.bubbles;
        this.type = event.type;
        this.timeStamp = event.timeStamp;
        this.id = crypto.randomUUID();
    }
}
export class FormSubmitExtensionEvent extends ExtensionEvent {
    targetValue;
    constructor(value, event) {
        super(event);
        this.targetValue = value;
    }
}
export const iconNames = {
    add: 'add',
    appointment: 'appointment',
    approvals: 'approvals',
    artificialIntelligence: 'artificialIntelligence',
    artificialIntelligenceEnhanced: 'artificialIntelligenceEnhanced',
    attach: 'attach',
    bank: 'bank',
    block: 'block',
    book: 'knowledgeBase',
    bulb: 'bulb',
    callTranscript: 'callTranscript',
    calling: 'calling',
    callingHangup: 'callingHangup',
    callingMade: 'callingMade',
    callingMissed: 'callingMissed',
    callingVoicemail: 'callingVoicemail',
    campaigns: 'campaigns',
    cap: 'cap',
    checkCircle: 'checkCircle',
    circleFilled: 'circleFilled',
    circleHollow: 'circleHollow',
    clock: 'time',
    comment: 'comments',
    contact: 'contacts',
    copy: 'duplicate',
    crm: 'crm',
    dataSync: 'dataSync',
    date: 'date',
    delay: 'delay',
    delete: 'delete',
    description: 'description',
    developerProjects: 'developerProjects',
    documents: 'documents',
    downCarat: 'downCarat',
    download: 'download',
    edit: 'edit',
    ellipses: 'ellipses',
    email: 'email',
    emailOpen: 'emailOpen',
    emailThreadedReplies: 'emailThreadedReplies',
    enrichment: 'enrichment',
    enroll: 'enroll',
    exclamation: 'exclamation',
    exclamationCircle: 'exclamationCircle',
    faceHappy: 'emoji',
    faceHappyFilled: 'emojiFillHappy',
    faceNeutral: 'emojiLineNeutral',
    faceNeutralFilled: 'emojiFillNeutral',
    faceSad: 'emojiLineSad',
    faceSadFilled: 'emojiFillSad',
    facebook: 'socialBlockFacebook',
    favoriteHollow: 'favoriteHollow',
    file: 'file',
    filledXCircleIcon: 'filledXCircleIcon',
    filter: 'filter',
    flame: 'highlyEngagedLead',
    folder: 'folder',
    folderOpen: 'folderOpen',
    forward: 'forward',
    gauge: 'gauge',
    generateChart: 'generateChart',
    gift: 'gift',
    globe: 'language',
    globeLine: 'globe',
    goal: 'goal',
    googlePlus: 'socialBlockGoogleplus',
    guidedActions: 'guidedActions',
    hash: 'numericDataType',
    hide: 'hide',
    home: 'home',
    hubDB: 'hubDB',
    image: 'insertImage',
    imageGallery: 'imageGallery',
    inbox: 'inbox',
    info: 'info',
    infoNoCircle: 'infoNoCircle',
    insertVideo: 'insertVideo',
    instagram: 'socialBlockInstagram',
    integrations: 'integrations',
    invoice: 'invoice',
    key: 'key',
    language: 'language',
    left: 'left',
    lessCircle: 'lessCircle',
    lesson: 'lesson',
    light: 'light',
    link: 'link',
    linkedin: 'socialBlockLinkedin',
    listView: 'listView',
    location: 'location',
    locked: 'locked',
    mention: 'mention',
    messages: 'messages',
    mobile: 'mobile',
    moreCircle: 'moreCircle',
    notEditable: 'notEditable',
    notification: 'notification',
    notificationOff: 'notificationOff',
    objectAssociations: 'objectAssociations',
    objectAssociationsManyToMany: 'objectAssociationsManyToMany',
    objectAssociationsManyToOne: 'objectAssociationsManyToOne',
    office365: 'office365',
    order: 'order',
    paymentSubscriptions: 'paymentSubscriptions',
    pin: 'pin',
    pinterest: 'socialBlockPinterest',
    powerPointFile: 'powerPointFile',
    presentation: 'presentation',
    product: 'product',
    publish: 'publish',
    question: 'question',
    questionAnswer: 'questionAnswer',
    questionCircle: 'questionCircle',
    quickbooks: 'quickbooks',
    quote: 'insertQuote',
    readMore: 'readMore',
    readOnlyView: 'readOnlyView',
    realEstateListing: 'realEstateListing',
    recentlySelected: 'recentlySelected',
    record: 'record',
    redo: 'redo',
    refresh: 'refresh',
    registration: 'registration',
    remove: 'remove',
    replace: 'replace',
    reports: 'reports',
    right: 'right',
    robot: 'simpleBot',
    rotate: 'rotate',
    rss: 'socialBlockRss',
    salesQuote: 'salesQuote',
    salesTemplates: 'salesTemplates',
    save: 'saveEditableView',
    search: 'search',
    send: 'send',
    sequences: 'sequences',
    settings: 'settings',
    shoppingCart: 'cart',
    signal: 'signal',
    signalPoor: 'signalPoor',
    signature: 'signature',
    snooze: 'snooze',
    sortAlpAsc: 'sortAlpAsc',
    sortAlpDesc: 'sortAlpDesc',
    sortAmtAsc: 'sortAmtAsc',
    sortAmtDesc: 'sortAmtDesc',
    sortNumAsc: 'sortNumAsc',
    sortNumDesc: 'sortNumDesc',
    sortTableAsc: 'sortTableAsc',
    sortTableDesc: 'sortTableDesc',
    spellCheck: 'spellCheck',
    sprocket: 'sprocket',
    star: 'favorite',
    stopRecord: 'stopRecord',
    strike: 'strike',
    styles: 'styles',
    success: 'success',
    tablet: 'tablet',
    tag: 'tag',
    tasks: 'tasks',
    test: 'test',
    text: 'text',
    textBodyExpanded: 'textBodyExpanded',
    textColor: 'textColor',
    textDataType: 'textDataType',
    textSnippet: 'textSnippet',
    thumbsDown: 'thumbsDown',
    thumbsUp: 'thumbsUp',
    ticket: 'ticket',
    translate: 'translate',
    trophy: 'trophy',
    twitter: 'socialBlockTwitter',
    undo: 'undo',
    upCarat: 'upCarat',
    upload: 'upload',
    video: 'video',
    videoFile: 'videoFile',
    videoPlayerSubtitles: 'videoPlayerSubtitles',
    view: 'view',
    viewDetails: 'viewDetails',
    warning: 'warning',
    website: 'website',
    workflows: 'workflows',
    x: 'socialBlockX',
    xCircle: 'xCircle',
    xing: 'socialBlockXing',
    youtube: 'socialBlockYoutube',
    youtubePlay: 'socialBlockYoutubeplay',
    zoomIn: 'zoomIn',
    zoomOut: 'zoomOut',
};
/** @deprecated use ExtensionEvent/FormSubmitExtensionEvent instead */
export class RemoteEvent {
    type;
    bubbles;
    timeStamp;
    targetValue;
    constructor(value, event) {
        this.bubbles = event.bubbles;
        this.type = event.type;
        this.timeStamp = event.timeStamp;
        this.targetValue = value;
    }
}
