// @hubspot/eslint-config-ui-extensions/dist/rules/no-browser-storage.js
import { isWindowSelfOrGlobalThisPropertyAccess } from "../utils/ast-utils.js";
import { isInTypeContext, isLocalVariable } from "../utils/scope-utils.js";
const STORAGE_IDENTIFIERS = new Set([
    'localStorage',
    'sessionStorage',
]);
export const noBrowserStorage = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Prevent usage of browser storage APIs (localStorage and sessionStorage) in UI Extensions, which run in a sandboxed web worker environment where these APIs are not available.',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-browser-storage',
        },
        messages: {
            noBrowserStorage: '{{ storageType }} is not available. Use React state or a server-side backend instead.',
        },
        schema: [],
    },
    create(context) {
        return {
            Identifier(node) {
                // Skip if not a storage identifier
                if (!STORAGE_IDENTIFIERS.has(node.name)) {
                    return;
                }
                const storageName = node.name;
                // Skip if in a TypeScript type context
                if (isInTypeContext(node)) {
                    return;
                }
                // Skip if this is a local variable
                if (isLocalVariable(context, node, storageName)) {
                    return;
                }
                // Skip if this identifier is the property being accessed (e.g., obj.localStorage)
                // We only want to catch direct storage usage, not property names
                const parent = node.parent;
                if (parent?.type === 'MemberExpression' &&
                    parent.property === node &&
                    !parent.computed) {
                    return;
                }
                // Skip if this identifier is a property key in an object literal
                // e.g., { localStorage: {} } - the key should be allowed
                if (parent?.type === 'Property' &&
                    parent.key === node &&
                    !parent.computed) {
                    // This is a property key in an object literal
                    return;
                }
                // This is a global storage access
                context.report({
                    node,
                    messageId: 'noBrowserStorage',
                    data: {
                        storageType: storageName,
                    },
                });
            },
            MemberExpression(node) {
                // Check for window.localStorage, self.localStorage, globalThis.localStorage
                // and the same for sessionStorage
                for (const storageName of STORAGE_IDENTIFIERS) {
                    if (isWindowSelfOrGlobalThisPropertyAccess(node, storageName)) {
                        context.report({
                            node,
                            messageId: 'noBrowserStorage',
                            data: {
                                storageType: storageName,
                            },
                        });
                        break;
                    }
                }
            },
        };
    },
};
