// @hubspot/eslint-config-ui-extensions/dist/rules/no-browser-dialogs.js
import { getMemberExpressionPropertyName, isGlobalIdentifierCall, isWindowSelfOrGlobalThisPropertyAccess, } from "../utils/ast-utils.js";
import { isInTypeContext, isLocalVariable } from "../utils/scope-utils.js";
const MESSAGE_IDS = {
    alert: 'nativeAlert',
    confirm: 'nativeConfirm',
    prompt: 'nativePrompt',
};
export const noBrowserDialogs = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Prevent usage of native browser dialog APIs (alert, confirm, and prompt) in UI Extensions.',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-browser-dialogs',
        },
        messages: {
            nativeAlert: 'alert() is not available. Use actions.addAlert() or <Alert> instead.',
            nativeConfirm: 'confirm() is not available. Use a <Modal> with <Button> instead.',
            nativePrompt: 'prompt() is not available. Use a <Modal> with <Input> instead.',
        },
        schema: [],
    },
    create(context) {
        return {
            CallExpression(node) {
                // Check for global dialog calls
                if (isGlobalIdentifierCall(node, 'alert') ||
                    isGlobalIdentifierCall(node, 'confirm') ||
                    isGlobalIdentifierCall(node, 'prompt')) {
                    const functionName = node.callee.name;
                    // Use scope analysis to check if the function is a local variable
                    if (isLocalVariable(context, node, functionName)) {
                        return; // Skip - this is a local variable, not the global function
                    }
                    // Skip if the function appears in a TypeScript type context
                    if (node.callee.type === 'Identifier' &&
                        isInTypeContext(node.callee)) {
                        return;
                    }
                    context.report({
                        node,
                        messageId: MESSAGE_IDS[functionName],
                    });
                    return;
                }
                if (isWindowSelfOrGlobalThisPropertyAccess(node, 'alert') ||
                    isWindowSelfOrGlobalThisPropertyAccess(node, 'confirm') ||
                    isWindowSelfOrGlobalThisPropertyAccess(node, 'prompt')) {
                    if (node.callee.type === 'MemberExpression') {
                        const propertyName = getMemberExpressionPropertyName(node.callee);
                        if (propertyName === undefined) {
                            return; // Skip - unable to determine property name
                        }
                        context.report({
                            node,
                            messageId: MESSAGE_IDS[propertyName],
                        });
                    }
                }
            },
        };
    },
};
