// @hubspot/eslint-config-ui-extensions/dist/rules/no-console.js
import { getMemberExpressionPropertyName } from "../utils/ast-utils.js";
import { isInTypeContext, isLocalVariable } from "../utils/scope-utils.js";
const CONSOLE_METHODS = ['log', 'info', 'warn', 'debug', 'error'];
const LOGGER_ALTERNATIVES = {
    log: 'logger.debug()',
    info: 'logger.info()',
    warn: 'logger.warn()',
    debug: 'logger.debug()',
    error: 'logger.error()',
};
export const noConsole = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Prevent usage of console methods in UI Extensions in favor of the SDK logger utility.',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-console',
        },
        messages: {
            noConsole: '{{ alternative }} is recommended instead of console.{{ method }}().',
        },
        schema: [],
    },
    create(context) {
        return {
            CallExpression(node) {
                const { callee } = node;
                // Check for console.method() calls (MemberExpression)
                if (callee.type === 'MemberExpression') {
                    const methodName = getMemberExpressionPropertyName(callee);
                    // Early exit if we can't determine the method name or it's not a console method
                    if (!methodName ||
                        !CONSOLE_METHODS.includes(methodName)) {
                        return;
                    }
                    const { object } = callee;
                    let isConsoleAccess = false;
                    // Check for direct console access: console.log
                    if (object.type === 'Identifier' && object.name === 'console') {
                        isConsoleAccess = true;
                    }
                    // Check for window.console.log, globalThis.console.log, self.console.log
                    if (object.type === 'MemberExpression') {
                        const outerObject = object.object;
                        const innerProperty = getMemberExpressionPropertyName(object);
                        if (outerObject.type === 'Identifier' &&
                            ['window', 'globalThis', 'self'].includes(outerObject.name) &&
                            innerProperty === 'console') {
                            isConsoleAccess = true;
                        }
                    }
                    // Early exit if not accessing console object
                    if (!isConsoleAccess) {
                        return;
                    }
                    // Skip if console is a local variable (shadowing the global)
                    if (object.type === 'Identifier' &&
                        object.name === 'console' &&
                        isLocalVariable(context, node, 'console')) {
                        return;
                    }
                    // Skip if console appears in a TypeScript type context
                    if (isInTypeContext(callee)) {
                        return;
                    }
                    // Report the violation
                    context.report({
                        node,
                        messageId: 'noConsole',
                        data: {
                            method: methodName,
                            alternative: LOGGER_ALTERNATIVES[methodName],
                        },
                    });
                }
            },
        };
    },
};
