// @hubspot/eslint-config-ui-extensions/dist/rules/no-dom-access.js
import { isWindowSelfOrGlobalThisPropertyAccess } from "../utils/ast-utils.js";
import { isInTypeContext, isLocalVariable } from "../utils/scope-utils.js";
export const noDomAccess = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Prevent access to the DOM (document object) in UI Extensions, which run in a sandboxed web worker without DOM access.',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-dom-access',
        },
        messages: {
            noDomAccess: 'document is not available. Use @hubspot/ui-extensions components instead.',
        },
        schema: [],
    },
    create(context) {
        return {
            Identifier(node) {
                // Skip if not the 'document' identifier
                if (node.name !== 'document') {
                    return;
                }
                // Skip if in a TypeScript type context
                if (isInTypeContext(node)) {
                    return;
                }
                // Skip if this is a local variable
                if (isLocalVariable(context, node, 'document')) {
                    return;
                }
                // Skip if this identifier is the property being accessed (e.g., obj.document)
                // We only want to catch direct 'document' usage, not property names
                const parent = node.parent;
                if (parent?.type === 'MemberExpression' &&
                    parent.property === node &&
                    !parent.computed) {
                    return;
                }
                // Skip if this identifier is a property key in an object literal
                // e.g., { document: {} } - the key should be allowed
                if (parent?.type === 'Property' &&
                    parent.key === node &&
                    !parent.computed) {
                    // This is a property key in an object literal
                    return;
                }
                // This is a global document access
                context.report({
                    node,
                    messageId: 'noDomAccess',
                });
            },
            MemberExpression(node) {
                // Check for window.document, self.document, or globalThis.document
                if (isWindowSelfOrGlobalThisPropertyAccess(node, 'document')) {
                    context.report({
                        node,
                        messageId: 'noDomAccess',
                    });
                }
            },
        };
    },
};
