// @hubspot/eslint-config-ui-extensions/dist/rules/no-experimental-imports.js
import { isExperimentalImport } from "../utils/import-utils.js";
export const noExperimentalImports = {
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Warn when importing from the experimental path of @hubspot/ui-extensions.',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-experimental-imports',
        },
        messages: {
            noExperimentalImports: 'Imports from "{{importSource}}" use experimental APIs. Experimental components are not stable and are not recommended for production.',
        },
        schema: [],
    },
    create(context) {
        return {
            ImportDeclaration(node) {
                const importSource = node.source.value;
                if (!isExperimentalImport(importSource)) {
                    return;
                }
                context.report({
                    node,
                    messageId: 'noExperimentalImports',
                    data: { importSource },
                });
            },
        };
    },
};
