// @hubspot/eslint-config-ui-extensions/dist/rules/no-html-elements.js
import { htmlTags } from "../constants.js";
export const unsupportedHtmlElements = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow usage of unsupported HTML elements in UI Extensions.',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-html-elements',
        },
        messages: {
            htmlNotSupported: '<{{ elementName }}> is not supported. Use @hubspot/ui-extensions components instead.',
        },
        schema: [],
    },
    create(context) {
        return {
            // A node of type JSXOpeningElement represents the opening tag of a JSX element.
            JSXOpeningElement(node) {
                /**
                 * JSXOpeningElement nodes have a 'name' property.
                 * The name can be a JSXIdentifier, JSXMemberExpression, or JSXNamespacedName.
                 * We are only interested in JSXIdentifier, which represents simple tag names like <div> or <MyComponent>
                 */
                if (node.name.type === 'JSXIdentifier') {
                    const element = node.name.name;
                    // If the element is a standard HTML tag, report it
                    if (htmlTags.has(element)) {
                        context.report({
                            node,
                            messageId: 'htmlNotSupported',
                            data: {
                                elementName: element,
                            },
                        });
                    }
                }
            },
        };
    },
};
