// @hubspot/eslint-config-ui-extensions/dist/rules/no-invalid-extension-point-imports.js
import path from 'path';
import { getExtensionPointFromFilePath, getHubspotImportExtensionPoint, mapExtensionPointToImportDirectory, } from "../utils/extension-point-utils.js";
import { isHubspotExtensibilityImport } from "../utils/import-utils.js";
export const noInvalidExtensionPointImports = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Prevent importing components from unsupported extension points.',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-invalid-extension-point-imports',
        },
        messages: {
            invalidExtensionPointImport: 'Imports from "{{importedPoint}}" are not allowed in {{currentPoint}}.',
        },
        schema: [],
    },
    create(context) {
        return {
            // A node of type ImportDeclaration represents an import statement in the code.
            ImportDeclaration(node) {
                // This is the value of the import source, e.g., 'my-module' in `import x from 'my-module';`
                const importSource = node.source.value;
                // If it's not a HubSpot UI Extensions import, we can skip it.
                if (!isHubspotExtensibilityImport(importSource)) {
                    return;
                }
                // Get the extension point from the import source
                const importExtensionPoint = getHubspotImportExtensionPoint(importSource);
                // Ignore imports that are not from recognized extension points.
                if (!importExtensionPoint) {
                    return;
                }
                // Determine the extension point from the file's path
                const filename = context.filename;
                const extensionPointDirectory = getExtensionPointFromFilePath(path.resolve(filename));
                if (!extensionPointDirectory) {
                    return;
                }
                // Get the mapped directory for the current extension point, to ensure our import paths + extension points align
                const mappedExtensionDirectory = mapExtensionPointToImportDirectory(extensionPointDirectory);
                // Finally, throw an error if the import extension point does not match the file's extension point
                if (importExtensionPoint !== mappedExtensionDirectory) {
                    context.report({
                        node,
                        messageId: 'invalidExtensionPointImport',
                        data: {
                            importedPoint: importExtensionPoint,
                            currentPoint: extensionPointDirectory,
                        },
                    });
                }
            },
        };
    },
};
