// @hubspot/eslint-config-ui-extensions/dist/rules/no-invalid-image-src.js
import { getImportedComponentName, isHubspotExtensibilityImport, } from "../utils/import-utils.js";
import { findJSXAttribute, isJSXComponent } from "../utils/jsx-utils.js";
const ALLOWED_PROTOCOLS = {
    HTTP: 'http:',
    HTTPS: 'https:',
    DATA: 'data:',
};
const allowedProtocols = Object.values(ALLOWED_PROTOCOLS);
export const noInvalidImageSrc = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Only allow valid image URLs in the src attribute of Image components from @hubspot/ui-extensions.',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-invalid-image-src',
        },
        messages: {
            invalidRelativeUrl: 'Image src must be a full URL (http://, https://, or data:), not a relative path.',
            whitespaceInUrl: 'Image src has leading or trailing whitespace. Remove it for the image to load.',
            invalidProtocol: 'Image src must use http://, https://, or data: protocol.',
            invalidImageType: 'Image src must point to a valid image file (.jpg, .png, .gif, .svg, .webp).',
        },
        schema: [],
        fixable: 'code',
    },
    create(context) {
        /**
         * This is used to track the name of the Image component being used in the file, if imported.
         * We can't assume it's always "Image" because it could be aliased on import.
         *
         * For example:
         * import { Image as HubImage } from '@hubspot/ui-extensions';
         *
         * In this case, we need to know that "HubImage" is the Image component, to properly lint its usage.
         */
        let imageComponentName = null;
        return {
            ImportDeclaration(node) {
                // Once we've found the Image component name, no need to keep looking.
                if (!imageComponentName) {
                    const importSource = node.source.value;
                    // If it's not a HubSpot UI Extensions import, skip it.
                    if (!isHubspotExtensibilityImport(importSource)) {
                        return;
                    }
                    // Otherwise, check for the Image component import and store its name.
                    imageComponentName = getImportedComponentName(node, 'Image');
                }
            },
            JSXOpeningElement(node) {
                // If we haven't identified the Image component in this file, skip it.
                if (!imageComponentName) {
                    return;
                }
                // If this JSX element is not the Image component, skip it.
                if (!isJSXComponent(node, imageComponentName)) {
                    return;
                }
                // Find the src attribute if it exists
                const srcAttribute = findJSXAttribute(node, 'src');
                if (!srcAttribute) {
                    return;
                }
                const srcAttributeValue = srcAttribute.value;
                // Only validate literal string values
                // JSX expressions and template literals are runtime values that we can't validate
                if (!srcAttributeValue ||
                    srcAttributeValue.type !== 'Literal' ||
                    typeof srcAttributeValue.value !== 'string') {
                    return;
                }
                const srcValue = srcAttributeValue.value;
                // 1. Check for whitespace first (with autofix)
                if (srcValue !== srcValue.trim()) {
                    context.report({
                        node: srcAttributeValue,
                        messageId: 'whitespaceInUrl',
                        fix(fixer) {
                            return fixer.replaceText(srcAttributeValue, `"${srcValue.trim()}"`);
                        },
                    });
                    return;
                }
                // 2. Try to parse the URL
                let url;
                try {
                    url = new URL(srcValue);
                }
                catch {
                    // Invalid or relative URL
                    context.report({
                        node: srcAttributeValue,
                        messageId: 'invalidRelativeUrl',
                    });
                    return;
                }
                // 3. Check protocol allowlist
                if (!allowedProtocols.includes(url.protocol)) {
                    context.report({
                        node: srcAttributeValue,
                        messageId: 'invalidProtocol',
                    });
                    return;
                }
                // 4. Check file extension (only for http/https URLs)
                if (url.protocol === ALLOWED_PROTOCOLS.HTTP ||
                    url.protocol === ALLOWED_PROTOCOLS.HTTPS) {
                    const validExtensions = /\.(jpg|jpeg|png|gif|svg|webp)$/i;
                    if (!validExtensions.test(url.pathname)) {
                        context.report({
                            node: srcAttributeValue,
                            messageId: 'invalidImageType',
                        });
                        return;
                    }
                }
                // If we get here, the URL is valid!
            },
        };
    },
};
