// @hubspot/eslint-config-ui-extensions/dist/rules/no-native-http.js
import { isGlobalIdentifierCall, isWindowSelfOrGlobalThisPropertyAccess, } from "../utils/ast-utils.js";
import { isInTypeContext, isLocalVariable } from "../utils/scope-utils.js";
export const noNativeHttp = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Prevent usage of native browser HTTP APIs (fetch and XMLHttpRequest) in UI Extensions.',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-native-http',
        },
        messages: {
            nativeFetch: 'Native fetch() is not available in UI Extensions. Use hubspot.fetch() instead.',
            nativeXMLHttpRequest: 'XMLHttpRequest is not available in UI Extensions. Use hubspot.fetch() instead.',
        },
        schema: [],
    },
    create(context) {
        return {
            CallExpression(node) {
                // Check for global fetch() call
                if (isGlobalIdentifierCall(node, 'fetch')) {
                    // Use scope analysis to check if 'fetch' is a local variable
                    // This handles cases like: const { fetch } = hubspot;
                    if (isLocalVariable(context, node, 'fetch')) {
                        return; // Skip - this is a local variable, not the global fetch
                    }
                    // Skip if fetch appears in a TypeScript type context
                    if (node.callee.type === 'Identifier' &&
                        isInTypeContext(node.callee)) {
                        return;
                    }
                    context.report({
                        node,
                        messageId: 'nativeFetch',
                    });
                    return;
                }
                // Check for window.fetch() or globalThis.fetch()
                if (isWindowSelfOrGlobalThisPropertyAccess(node, 'fetch')) {
                    context.report({
                        node,
                        messageId: 'nativeFetch',
                    });
                }
            },
            /**
             * Detect new XMLHttpRequest() constructor calls
             */
            NewExpression(node) {
                // Early exit: callee is not an Identifier
                if (node.callee.type !== 'Identifier') {
                    return;
                }
                if (node.callee.name === 'XMLHttpRequest') {
                    // Use scope analysis to check if 'XMLHttpRequest' is a local variable
                    if (isLocalVariable(context, node, 'XMLHttpRequest')) {
                        return; // Skip - this is a local variable or import
                    }
                    // Skip if XMLHttpRequest appears in a TypeScript type context
                    if (isInTypeContext(node.callee)) {
                        return;
                    }
                    context.report({
                        node,
                        messageId: 'nativeXMLHttpRequest',
                    });
                }
            },
        };
    },
};
