// @hubspot/eslint-config-ui-extensions/dist/rules/no-parent-imports.js
import { getExtensionPointFromFilePath, isImportOutsideExtensionPoint, } from "../utils/extension-point-utils.js";
import { isRelativeImport } from "../utils/import-utils.js";
export const noParentImports = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Prevent importing files from outside the extension point root directory',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-parent-imports',
        },
        messages: {
            noParentImports: 'Imports from outside "{{ extensionPoint }}" will fail at build time. Use npm workspaces to share code.',
        },
        schema: [],
    },
    create(context) {
        return {
            // A node of type ImportDeclaration represents an import statement in the code.
            ImportDeclaration(node) {
                // This is the value of the import source, e.g., './utils' in `import x from './utils';`
                const importSource = node.source.value;
                // Only check relative imports (starting with ./ or ../)
                if (!isRelativeImport(importSource)) {
                    return;
                }
                // Get the current file's path
                const filename = context.filename;
                // Determine the extension point from the file's path
                const extensionPoint = getExtensionPointFromFilePath(filename);
                // Rule only applies to files in extension point directories
                if (!extensionPoint) {
                    return;
                }
                // Check if the import escapes the extension point boundary
                if (isImportOutsideExtensionPoint(filename, importSource)) {
                    context.report({
                        node,
                        messageId: 'noParentImports',
                        data: {
                            extensionPoint,
                        },
                    });
                }
            },
        };
    },
};
