// @hubspot/eslint-config-ui-extensions/dist/rules/no-restricted-globals.js
import { isGlobalIdentifierCall, isWindowSelfOrGlobalThisPropertyAccess, } from "../utils/ast-utils.js";
import { isInTypeContext, isLocalVariable } from "../utils/scope-utils.js";
// Function-like globals that should only be checked as calls
const RESTRICTED_FUNCTION_GLOBALS = new Set([
    'importScripts',
    'addEventListener',
    'removeEventListener',
]);
// Non-function globals that can be accessed as identifiers or members
const RESTRICTED_VALUE_GLOBALS = new Set([
    'location',
    'history',
    'navigator',
]);
// All restricted globals (for Identifier checking)
const ALL_RESTRICTED_GLOBALS = new Set([
    ...RESTRICTED_FUNCTION_GLOBALS,
    ...RESTRICTED_VALUE_GLOBALS,
]);
export const noRestrictedGlobals = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Prevent usage of browser globals that are not available in the sandboxed web worker environment.',
            url: 'https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/linting/rules/no-restricted-globals',
        },
        messages: {
            importScripts: 'importScripts() is not supported. Use ES modules (import/export) instead.',
            addEventListener: 'addEventListener() is not supported. Use React event props (onClick, etc.) instead.',
            removeEventListener: 'removeEventListener() is not supported. Use React event props instead.',
            location: 'location is not available. Use context.location instead.',
            history: 'history is not available. Use <Link> or actions.openIframeModal() instead.',
            navigator: 'navigator is not available as expected. Use provided context and actions instead.',
        },
        schema: [],
    },
    create(context) {
        return {
            Identifier(node) {
                // Skip if not a restricted global
                if (!ALL_RESTRICTED_GLOBALS.has(node.name)) {
                    return;
                }
                const globalName = node.name;
                // Skip if in a TypeScript type context
                if (isInTypeContext(node)) {
                    return;
                }
                // Skip if this is a local variable
                if (isLocalVariable(context, node, globalName)) {
                    return;
                }
                // Skip if this identifier is the property being accessed (e.g., obj.location)
                // We only want to catch direct usage, not property names
                const parent = node.parent;
                if (parent?.type === 'MemberExpression' &&
                    parent.property === node &&
                    !parent.computed) {
                    return;
                }
                // Skip if this identifier is a property key in an object literal
                // e.g., { location: {} } - the key should be allowed
                if (parent?.type === 'Property' &&
                    parent.key === node &&
                    !parent.computed) {
                    return;
                }
                // Skip if this identifier is a property key in a TypeScript interface/type
                // e.g., interface Router { history: typeof history; } - the first 'history' should be allowed
                // TSPropertySignature is a TS-parser node type not in ESTree's type definitions
                if (parent &&
                    parent.type === 'TSPropertySignature' &&
                    'key' in parent &&
                    parent.key === node &&
                    'computed' in parent &&
                    !parent.computed) {
                    return;
                }
                // Skip if this identifier is the callee of a CallExpression for function-like globals
                // The CallExpression handler will catch these cases
                if (parent?.type === 'CallExpression' &&
                    parent.callee === node &&
                    RESTRICTED_FUNCTION_GLOBALS.has(globalName)) {
                    return;
                }
                // This is a global access
                context.report({
                    node,
                    messageId: globalName,
                });
            },
            MemberExpression(node) {
                // Check for window.location, self.history, etc.
                // Only check value globals (functions are handled by CallExpression)
                for (const globalName of RESTRICTED_VALUE_GLOBALS) {
                    if (isWindowSelfOrGlobalThisPropertyAccess(node, globalName)) {
                        context.report({
                            node,
                            messageId: globalName,
                        });
                        break;
                    }
                }
            },
            CallExpression(node) {
                // Check for function calls of restricted function globals
                for (const globalName of RESTRICTED_FUNCTION_GLOBALS) {
                    if (isGlobalIdentifierCall(node, globalName)) {
                        if (isLocalVariable(context, node, globalName)) {
                            return;
                        }
                        if (node.callee.type === 'Identifier' &&
                            isInTypeContext(node.callee)) {
                            return;
                        }
                        context.report({
                            node,
                            messageId: globalName,
                        });
                        return;
                    }
                    // Also check for window.importScripts(), self.addEventListener(), etc.
                    if (isWindowSelfOrGlobalThisPropertyAccess(node, globalName)) {
                        context.report({
                            node,
                            messageId: globalName,
                        });
                        return;
                    }
                }
            },
        };
    },
};
