// @hubspot/api-client/lib/codegen/crm/objects/deal_splits/types/PromiseAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromiseBatchApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class PromiseBatchApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservableBatchApi(configuration, requestFactory, responseProcessor);
    }
    readWithHttpInfo(batchInputPublicObjectId, _options) {
        const result = this.api.readWithHttpInfo(batchInputPublicObjectId, _options);
        return result.toPromise();
    }
    read(batchInputPublicObjectId, _options) {
        const result = this.api.read(batchInputPublicObjectId, _options);
        return result.toPromise();
    }
    upsertWithHttpInfo(publicDealSplitsBatchCreateRequest, _options) {
        const result = this.api.upsertWithHttpInfo(publicDealSplitsBatchCreateRequest, _options);
        return result.toPromise();
    }
    upsert(publicDealSplitsBatchCreateRequest, _options) {
        const result = this.api.upsert(publicDealSplitsBatchCreateRequest, _options);
        return result.toPromise();
    }
}
exports.PromiseBatchApi = PromiseBatchApi;
