// @hubspot/ui-extensions/dist/shared/types/components/illustration.js
export const illustrationNames = {
    // Standard 1:1 mappings
    lock: 'lock',
    unlock: 'unlock',
    electronicSignature: 'electronic-signature',
    electronicSignatureEmptyState: 'electronic-signature-empty-state',
    callingSetUp: 'calling-set-up',
    companies: 'companies',
    contacts: 'contacts',
    multipleObjects: 'multiple-objects',
    object: 'object',
    task: 'task',
    deals: 'deals',
    registration: 'registration',
    developerSecurityUpdate: 'developer-security-update',
    social: 'social',
    leads: 'leads',
    target: 'target',
    emailConfirmation: 'email-confirmation',
    successfullyConnectedEmail: 'successfully-connected-email',
    contentStrategy: 'content-strategy',
    addOnReporting: 'add-on-reporting',
    productsShoppingCart: 'products-shopping-cart',
    store: 'store',
    storeDisabled: 'store-disabled',
    customerSupport: 'customer-support',
    customerExperience: 'customer-experience',
    api: 'api',
    integrations: 'integrations',
    sandboxAddOn: 'sandbox-add-on',
    customObjects: 'custom-objects',
    paymentsButton: 'payments-button',
    automatedTesting: 'automated-testing',
    emptyStateCharts: 'empty-state-charts',
    // Special mappings that are not 1:1
    meetings: 'meetings-small',
    tickets: 'ticket-small',
    errorGeneral: 'errors/general',
    errorHourglass: 'errors/hourglass',
    middlePaneCards: 'settings-middle-pane-cards',
    propertiesSidebar: 'settings-properties-sidebar',
    sidebar: 'settings-sidebar',
};
