// @hubspot/ui-extensions/dist/hooks/useCrmSearch.js
import { useCallback, useReducer, useState } from 'react';
import { useStableValue } from "../internal/hook-utils.js";
import { calculatePaginationFlags, DEFAULT_PAGE_SIZE, } from "../utils/pagination.js";
import { fetchCrmSearch } from "./utils/fetchCrmSearch.js";
import { useFetchLifecycle } from "./utils/useFetchLifecycle.js";
function createInitialState() {
    return {
        results: [],
        total: 0,
        error: null,
        isLoading: true,
        isRefetching: false,
        currentPage: 1,
        hasMore: false,
        currentCursor: undefined,
        nextCursor: undefined,
        offsetHistory: [],
    };
}
function crmSearchReducer(state, action) {
    switch (action.type) {
        case 'FETCH_START':
            return {
                ...state,
                isLoading: true,
                error: null,
            };
        case 'FETCH_SUCCESS':
            return {
                ...state,
                isLoading: false,
                results: action.payload.results,
                total: action.payload.total,
                hasMore: action.payload.hasMore,
                currentCursor: action.payload.currentCursor,
                nextCursor: action.payload.nextCursor,
                error: null,
            };
        case 'FETCH_ERROR':
            return {
                ...state,
                isLoading: false,
                error: action.payload,
                results: [],
                total: 0,
                hasMore: false,
                currentCursor: undefined,
                nextCursor: undefined,
            };
        case 'NEXT_PAGE':
            return {
                ...state,
                currentPage: state.currentPage + 1,
                offsetHistory: state.currentCursor !== undefined
                    ? [...state.offsetHistory, state.currentCursor]
                    : state.offsetHistory,
                currentCursor: state.nextCursor,
                nextCursor: undefined,
            };
        case 'PREVIOUS_PAGE': {
            const newPage = Math.max(1, state.currentPage - 1);
            const newHistory = [...state.offsetHistory];
            const previousCursor = newHistory.pop();
            return {
                ...state,
                currentPage: newPage,
                offsetHistory: newHistory,
                currentCursor: previousCursor,
                nextCursor: undefined,
            };
        }
        case 'RESET':
            return {
                ...state,
                currentPage: 1,
                results: [],
                total: 0,
                isLoading: true,
                hasMore: false,
                error: null,
                currentCursor: undefined,
                nextCursor: undefined,
                offsetHistory: [],
            };
        case 'REFETCH_START':
            return {
                ...state,
                isRefetching: true,
                error: null,
            };
        case 'REFETCH_SUCCESS':
            return {
                ...state,
                isRefetching: false,
                results: action.payload.results,
                total: action.payload.total,
                hasMore: action.payload.hasMore,
                nextCursor: action.payload.nextCursor,
                error: null,
            };
        case 'REFETCH_ERROR':
            return {
                ...state,
                isRefetching: false,
                error: action.payload,
            };
        default:
            return state;
    }
}
const DEFAULT_OPTIONS = {};
export function useCrmSearch(config, options = DEFAULT_OPTIONS) {
    const pageSize = config?.pageLength ?? DEFAULT_PAGE_SIZE;
    const [state, dispatch] = useReducer(crmSearchReducer, createInitialState());
    // Reset pagination when pageLength changes.
    // Render-phase update (not useEffect) avoids committing a render with stale paginated results.
    const [prevPageSize, setPrevPageSize] = useState(pageSize);
    if (prevPageSize !== pageSize) {
        setPrevPageSize(pageSize);
        dispatch({ type: 'RESET' });
    }
    const stableConfig = useStableValue(config);
    const stableOptions = useStableValue(options);
    const nextPage = useCallback(() => {
        const paginationFlags = calculatePaginationFlags(state.currentPage, state.hasMore);
        if (paginationFlags.hasNextPage) {
            dispatch({ type: 'NEXT_PAGE' });
        }
    }, [state.currentPage, state.hasMore]);
    const previousPage = useCallback(() => {
        const paginationFlags = calculatePaginationFlags(state.currentPage, state.hasMore);
        if (paginationFlags.hasPreviousPage) {
            dispatch({ type: 'PREVIOUS_PAGE' });
        }
    }, [state.currentPage, state.hasMore]);
    const reset = useCallback(() => {
        const paginationFlags = calculatePaginationFlags(state.currentPage, state.hasMore);
        if (paginationFlags.hasPreviousPage) {
            dispatch({ type: 'RESET' });
        }
    }, [state.currentPage, state.hasMore]);
    const { refetch } = useFetchLifecycle({
        fetchFn: () => {
            const request = {
                objectType: stableConfig?.objectType,
                properties: stableConfig?.properties,
                query: stableConfig?.query,
                filterGroups: stableConfig?.filterGroups,
                sorts: stableConfig?.sorts,
                pageLength: pageSize,
                after: state.currentCursor,
            };
            return fetchCrmSearch(request, {
                propertiesToFormat: stableOptions.propertiesToFormat,
                formattingOptions: stableOptions.formattingOptions,
            });
        },
        callbacks: {
            onStart: ({ isRefetch }) => dispatch({ type: isRefetch ? 'REFETCH_START' : 'FETCH_START' }),
            onSuccess: (data, { isRefetch }) => dispatch(isRefetch
                ? {
                    type: 'REFETCH_SUCCESS',
                    payload: {
                        results: data.results,
                        total: data.total,
                        hasMore: data.hasMore,
                        nextCursor: data.after,
                    },
                }
                : {
                    type: 'FETCH_SUCCESS',
                    payload: {
                        results: data.results,
                        total: data.total,
                        hasMore: data.hasMore,
                        nextCursor: data.after,
                        currentCursor: state.currentCursor,
                    },
                }),
            onError: (error, { isRefetch }) => dispatch({
                type: isRefetch ? 'REFETCH_ERROR' : 'FETCH_ERROR',
                payload: error,
            }),
        },
        deps: [stableConfig, stableOptions, state.currentCursor],
        defaultErrorMessage: 'Failed to fetch CRM search results',
    });
    const paginationFlags = calculatePaginationFlags(state.currentPage, state.hasMore);
    return {
        results: state.results,
        total: state.total,
        error: state.error,
        isLoading: state.isLoading,
        isRefetching: state.isRefetching,
        refetch,
        pagination: {
            hasNextPage: paginationFlags.hasNextPage,
            hasPreviousPage: paginationFlags.hasPreviousPage,
            currentPage: state.currentPage,
            pageSize,
            nextPage,
            previousPage,
            reset,
        },
    };
}
