// @hubspot/eslint-config-ui-extensions/dist/utils/extension-point-utils.js
import path from 'path';
import { EXTENSION_IMPORT_DIRECTORIES, EXTENSION_POINT_DIRECTORIES, } from "../constants.js";
/**
 * This function takes a file path as input and returns the extension point.
 *
 * It searches for the 'src/app/<extension-point>' pattern in the path,
 * where <extension-point> is one of the valid extension points.
 *
 * If the pattern is found, it returns the extension point name; otherwise, it returns null.
 */
export const getExtensionPointFromFilePath = (filePath) => {
    const segments = filePath.split(path.sep);
    for (let i = 0; i < segments.length - 2; i++) {
        if (segments[i] === 'src' && segments[i + 1] === 'app') {
            if (EXTENSION_POINT_DIRECTORIES.has(segments[i + 2])) {
                return segments[i + 2];
            }
        }
    }
    return null;
};
/**
 * This function extracts the extension point from the import source of hubspot components, hooks, etc.
 *
 * In our package, the import can come from different extension points.
 * The extension point is always in the third segment of the import path.
 *
 * For example:
 * import { useCrmProperties } from '@hubspot/ui-extensions/crm/hooks';
 *
 * It ignores non-extension-point imports, including those that import from other directories such as:
 *
 * import { useExtensionContext } from '@hubspot/ui-extensions/hooks';
 *
 * If the import source does not correspond to a known extension point, it returns null.
 */
export const getHubspotImportExtensionPoint = (importSource) => {
    const importParts = importSource.split('/');
    if (importParts.length > 2 &&
        EXTENSION_IMPORT_DIRECTORIES.has(importParts[2])) {
        return importParts[2];
    }
    return null;
};
/**
 * There are some extension points that map to different directory names.
 * For example, both 'extensions' and 'cards' map to the 'crm' directory.
 *
 * This function takes an extension point and returns the corresponding directory name.
 * If there is no special mapping, it returns the extension point as is.
 */
export const mapExtensionPointToImportDirectory = (extensionPoint) => {
    if (!extensionPoint) {
        return null;
    }
    // Map 'extensions' and 'cards' to 'crm'
    if (extensionPoint === 'extensions' || extensionPoint === 'cards') {
        return 'crm';
    }
    return extensionPoint;
};
/**
 * Checks if a relative import escapes the extension point directory boundary.
 *
 * Extension points (cards, settings, pages, extensions) must be self-contained.
 * While relative imports that go outside the extension point directory work locally,
 * they fail when the project is built and uploaded, as each extension point is
 * deployed as an independent unit.
 *
 * This function:
 * 1. Identifies the extension point from the file path
 * 2. Finds the extension point root directory
 * 3. Resolves the relative import path
 * 4. Checks if the resolved path is outside the extension point root
 *
 * For example, given file: /project/src/app/cards/Component.tsx
 * - import '../shared/utils' → returns true (escapes cards directory)
 * - import './utils' → returns false (stays within cards directory)
 * - import '../nested/Component' → returns false (stays within cards directory)
 *
 * @param currentFilePath - The absolute path of the file containing the import
 * @param importSource - The relative import source string
 * @returns True if the import goes outside the extension point boundary
 */
export const isImportOutsideExtensionPoint = (currentFilePath, importSource) => {
    // Get the extension point from the file path
    const extensionPoint = getExtensionPointFromFilePath(currentFilePath);
    if (!extensionPoint) {
        // Not in an extension point, rule doesn't apply
        return false;
    }
    // Find the extension point root directory using string matching
    const pattern = `${path.sep}src${path.sep}app${path.sep}${extensionPoint}`;
    const patternIndex = currentFilePath.indexOf(pattern);
    if (patternIndex === -1) {
        // Shouldn't happen if getExtensionPointFromFilePath returned a value
        return false;
    }
    // Construct the extension point root path
    const extensionPointRoot = currentFilePath.substring(0, patternIndex + pattern.length);
    // Resolve the import path relative to the current file
    const currentDir = path.dirname(currentFilePath);
    const resolvedImportPath = path.resolve(currentDir, importSource);
    // Use path.relative to check if the import escapes the extension point boundary
    // If the relative path starts with '..', it means the resolved path is outside
    const relativePath = path.relative(extensionPointRoot, resolvedImportPath);
    return relativePath.startsWith('..') || path.isAbsolute(relativePath);
};
