// @hubspot/ui-extensions/dist/crm/hooks/useAssociations.js
import { useCallback, useReducer, useState } from 'react';
import { useFetchLifecycle } from "../../hooks/utils/useFetchLifecycle.js";
import { createMockAwareHook, useStableValue, } from "../../internal/hook-utils.js";
import { calculatePaginationFlags, DEFAULT_PAGE_SIZE, } from "../../utils/pagination.js";
import { fetchAssociations } from "../utils/fetchAssociations.js";
function createInitialState() {
    return {
        results: [],
        error: null,
        isLoading: true,
        isRefetching: false,
        currentPage: 1,
        hasMore: false,
        currentOffset: undefined,
        nextOffset: undefined,
        offsetHistory: [],
    };
}
function associationsReducer(state, action) {
    switch (action.type) {
        case 'FETCH_START':
            return {
                ...state,
                isLoading: true,
                error: null,
            };
        case 'FETCH_SUCCESS':
            return {
                ...state,
                isLoading: false,
                results: action.payload.results,
                hasMore: action.payload.hasMore,
                currentOffset: action.payload.currentOffset,
                nextOffset: action.payload.nextOffset,
                error: null,
            };
        case 'FETCH_ERROR':
            return {
                ...state,
                isLoading: false,
                error: action.payload,
                results: [],
                hasMore: false,
                currentOffset: undefined,
                nextOffset: undefined,
            };
        case 'NEXT_PAGE':
            return {
                ...state,
                currentPage: state.currentPage + 1,
                offsetHistory: state.currentOffset !== undefined
                    ? [...state.offsetHistory, state.currentOffset]
                    : state.offsetHistory,
                currentOffset: state.nextOffset,
                nextOffset: undefined,
            };
        case 'PREVIOUS_PAGE': {
            const newPage = Math.max(1, state.currentPage - 1);
            const newHistory = [...state.offsetHistory];
            const previousOffset = newHistory.pop();
            return {
                ...state,
                currentPage: newPage,
                offsetHistory: newHistory,
                currentOffset: previousOffset,
                nextOffset: undefined,
            };
        }
        case 'RESET':
            return {
                ...state,
                currentPage: 1,
                results: [],
                isLoading: true,
                hasMore: false,
                error: null,
                currentOffset: undefined,
                nextOffset: undefined,
                offsetHistory: [],
            };
        case 'REFETCH_START':
            return {
                ...state,
                isRefetching: true,
                error: null,
            };
        case 'REFETCH_SUCCESS':
            return {
                ...state,
                isRefetching: false,
                results: action.payload.results,
                hasMore: action.payload.hasMore,
                nextOffset: action.payload.nextOffset,
                error: null,
            };
        case 'REFETCH_ERROR':
            return {
                ...state,
                isRefetching: false,
                error: action.payload,
            };
        default:
            return state;
    }
}
const DEFAULT_OPTIONS = {};
/**
 * A hook to fetch and manage associations between CRM objects with pagination support.
 */
function useAssociationsInternal(config, options = DEFAULT_OPTIONS) {
    const pageSize = config?.pageLength ?? DEFAULT_PAGE_SIZE;
    const [state, dispatch] = useReducer(associationsReducer, createInitialState());
    /**
     * Reset pagination when pageSize changes.
     * Render-phase update (not useEffect) avoids committing a render with stale paginated results.
     */
    const [prevPageSize, setPrevPageSize] = useState(pageSize);
    if (prevPageSize !== pageSize) {
        setPrevPageSize(pageSize);
        dispatch({ type: 'RESET' });
    }
    const stableConfig = useStableValue(config);
    const stableOptions = useStableValue(options);
    // Pagination actions
    const nextPage = useCallback(() => {
        const paginationFlags = calculatePaginationFlags(state.currentPage, state.hasMore);
        if (paginationFlags.hasNextPage) {
            dispatch({ type: 'NEXT_PAGE' });
        }
    }, [state.currentPage, state.hasMore]);
    const previousPage = useCallback(() => {
        const paginationFlags = calculatePaginationFlags(state.currentPage, state.hasMore);
        if (paginationFlags.hasPreviousPage) {
            dispatch({ type: 'PREVIOUS_PAGE' });
        }
    }, [state.currentPage, state.hasMore]);
    const reset = useCallback(() => {
        const paginationFlags = calculatePaginationFlags(state.currentPage, state.hasMore);
        // Only reset to first page if we're not already on the first page
        if (paginationFlags.hasPreviousPage) {
            dispatch({ type: 'RESET' });
        }
    }, [state.currentPage, state.hasMore]);
    const { refetch } = useFetchLifecycle({
        fetchFn: () => {
            const request = {
                toObjectType: stableConfig?.toObjectType,
                properties: stableConfig?.properties,
                pageLength: pageSize,
                offset: state.currentOffset,
            };
            return fetchAssociations(request, {
                propertiesToFormat: stableOptions.propertiesToFormat,
                formattingOptions: stableOptions.formattingOptions,
            });
        },
        callbacks: {
            onStart: ({ isRefetch }) => dispatch({ type: isRefetch ? 'REFETCH_START' : 'FETCH_START' }),
            onSuccess: (data, { isRefetch }) => dispatch(isRefetch
                ? {
                    type: 'REFETCH_SUCCESS',
                    payload: {
                        results: data.results,
                        hasMore: data.hasMore,
                        nextOffset: data.nextOffset,
                    },
                }
                : {
                    type: 'FETCH_SUCCESS',
                    payload: {
                        results: data.results,
                        hasMore: data.hasMore,
                        nextOffset: data.nextOffset,
                        currentOffset: state.currentOffset,
                    },
                }),
            onError: (error, { isRefetch }) => dispatch({
                type: isRefetch ? 'REFETCH_ERROR' : 'FETCH_ERROR',
                payload: error,
            }),
        },
        deps: [
            stableConfig,
            stableOptions,
            state.currentPage,
            state.currentOffset,
            pageSize,
        ],
        defaultErrorMessage: 'Failed to fetch associations',
    });
    // Calculate pagination flags
    const paginationFlags = calculatePaginationFlags(state.currentPage, state.hasMore);
    return {
        results: state.results,
        error: state.error,
        isLoading: state.isLoading,
        isRefetching: state.isRefetching,
        refetch,
        pagination: {
            hasNextPage: paginationFlags.hasNextPage,
            hasPreviousPage: paginationFlags.hasPreviousPage,
            currentPage: state.currentPage,
            pageSize,
            nextPage,
            previousPage,
            reset,
        },
    };
}
export const useAssociations = createMockAwareHook('useAssociations', useAssociationsInternal);
