// @hubspot/ui-extensions/dist/crm/hooks/useCrmProperties.js
import { useMemo, useReducer } from 'react';
import { useFetchLifecycle } from "../../hooks/utils/useFetchLifecycle.js";
import { createMockAwareHook, useStableValue, } from "../../internal/hook-utils.js";
import { fetchCrmProperties } from "../utils/fetchCrmProperties.js";
const initialState = {
    properties: {},
    error: null,
    isLoading: true,
    isRefetching: false,
};
function crmPropertiesReducer(state, action) {
    switch (action.type) {
        case 'FETCH_START':
            return {
                ...state,
                isLoading: true,
                error: null,
            };
        case 'FETCH_SUCCESS':
            return {
                ...state,
                isLoading: false,
                properties: action.payload,
                error: null,
            };
        case 'FETCH_ERROR':
            return {
                ...state,
                isLoading: false,
                error: action.payload,
                properties: {},
            };
        case 'REFETCH_START':
            return {
                ...state,
                isRefetching: true,
                error: null,
            };
        case 'REFETCH_SUCCESS':
            return {
                ...state,
                isRefetching: false,
                properties: action.payload,
                error: null,
            };
        case 'REFETCH_ERROR':
            return {
                ...state,
                isRefetching: false,
                error: action.payload,
            };
        default:
            return state;
    }
}
const DEFAULT_OPTIONS = {};
/**
 * A hook for using and managing CRM properties.
 */
function useCrmPropertiesInternal(propertyNames, options = DEFAULT_OPTIONS) {
    const [state, dispatch] = useReducer(crmPropertiesReducer, initialState);
    // Sort property names so that order doesn't matter for cache keys
    const sortedPropertyNames = useMemo(() => Array.isArray(propertyNames) ? [...propertyNames].sort() : propertyNames, [propertyNames]);
    const stablePropertyNames = useStableValue(sortedPropertyNames);
    const stableOptions = useStableValue(options);
    const { refetch } = useFetchLifecycle({
        fetchFn: (signal) => fetchCrmProperties(stablePropertyNames, (data) => {
            if (!signal.cancelled) {
                dispatch({
                    type: signal.isRefetch ? 'REFETCH_SUCCESS' : 'FETCH_SUCCESS',
                    payload: data,
                });
            }
        }, stableOptions),
        callbacks: {
            onStart: ({ isRefetch }) => dispatch({ type: isRefetch ? 'REFETCH_START' : 'FETCH_START' }),
            onSuccess: (data, { isRefetch }) => dispatch({
                type: isRefetch ? 'REFETCH_SUCCESS' : 'FETCH_SUCCESS',
                payload: data,
            }),
            onError: (error, { isRefetch }) => dispatch({
                type: isRefetch ? 'REFETCH_ERROR' : 'FETCH_ERROR',
                payload: error,
            }),
        },
        deps: [stablePropertyNames, stableOptions],
        defaultErrorMessage: 'Failed to fetch CRM properties',
    });
    return {
        ...state,
        refetch,
    };
}
export const useCrmProperties = createMockAwareHook('useCrmProperties', useCrmPropertiesInternal);
