// @hubspot/ui-extensions/dist/internal/global-utils.js
/**
 * Checks if the current environment is a HubSpot extension worker.
 * @returns True if the current environment is a HubSpot extension worker.
 */
const isRunningInWorker = () => typeof self !== 'undefined' &&
    self.__HUBSPOT_EXTENSION_WORKER__ === true;
/**
 * A fake worker globals object for use in test environments.
 */
const fakeWorkerGlobals = {
    logger: {
        debug: (data) => {
            console.log(data);
        },
        info: (data) => {
            console.info(data);
        },
        warn: (data) => {
            console.warn(data);
        },
        error: (data) => {
            console.error(data);
        },
    },
    extend_V2: () => {
        // No-op in test environment
    },
    // @ts-expect-error we are not using the worker endpoint in tests env.
    __useExtensionContext: () => {
        // No-op in test environment
    },
};
/**
 * Gets the worker globals object for the current environment.
 * @returns The worker globals object.
 */
export const getWorkerGlobals = () => {
    return isRunningInWorker()
        ? self
        : fakeWorkerGlobals;
};
