// @hubspot/ui-extensions/dist/internal/hook-utils.js
import { createContext, useContext, useRef } from 'react';
const ReactRenderMocksContext = createContext(null);
/**
 * Creates a mock-aware hook function that can be used to mock the original hook function.
 * The mock-aware hook function will return the mocked hook function if a mock is found, otherwise it will return the original hook function.
 *
 * @param hookName The name of the hook to mock that corresponds to the key in the Mocks interface
 * @param originalHookFunction The original hook function to call if no mock is found
 * @returns The mocked hook function or the original hook function if no mock is found
 */
export const createMockAwareHook = (hookName, originalHookFunction) => {
    const useWrapper = (...args) => {
        const mocksContext = useMocksContext();
        if (!mocksContext) {
            // If no mocks are provided, call the original hook function
            return originalHookFunction(...args);
        }
        const { mocks } = mocksContext;
        const mockHook = mocks[hookName];
        if (!mockHook) {
            throw new Error(`Illegal State: Mock for hook ${hookName} not found.`);
        }
        return mockHook(...args);
    };
    return useWrapper;
};
/**
 * A hook that provides access to the Mocks context.
 * Returns the mocks context value if inside a MocksContextProvider, otherwise returns null.
 *
 * @returns The mocks context value or null if not in a test environment.
 */
export function useMocksContext() {
    return useContext(ReactRenderMocksContext);
}
/**
 * A React component that provides the Mocks context that can be used to provide mocks to the mock-aware hook functions.
 *
 * @param children The children to render.
 * @returns The children wrapped in the Mocks context provider.
 */
export const MocksContextProvider = ReactRenderMocksContext.Provider;
/**
 * Stabilizes a value's reference identity across re-renders using deep
 * comparison via JSON.stringify.
 */
export function useStableValue(value) {
    const stableRef = useRef({
        key: JSON.stringify(value),
        value,
    });
    const key = JSON.stringify(value);
    if (key !== stableRef.current.key) {
        stableRef.current = { key, value };
    }
    return stableRef.current.value;
}
