// @hubspot/ui-extensions/dist/testing/internal/element.js
import { RenderedNodeType } from "../types.js";
import { isMatch } from "../utils.js";
import { EMPTY_RENDERED_CHILD_NODES } from "./constants.js";
import { debugLog } from "./debug.js";
import { InvalidEventFunctionError, MissingEventFunctionError, } from "./errors.js";
import { printNode } from "./print.js";
import { find, findAll, findAllChildren, findChild, maybeFind, maybeFindChild, } from "./query.js";
import { asRenderedElementNode, asRenderedNode, } from "./type-utils-internal.js";
/**
 * Creates a rendered element node.
 *
 * @param document The document that the element belongs to.
 * @param name The name of the element.
 * @param props The props of the element.
 * @returns A rendered element node.
 */
export const createElementNode = (document, name, props) => {
    const elementNode = {
        nodeType: RenderedNodeType.Element,
        name,
        props,
        text: null,
        childNodes: EMPTY_RENDERED_CHILD_NODES,
        document,
        debugLog: (label) => {
            debugLog(asRenderedNode(elementNode), label);
        },
        find: (targetComponent, matcher) => {
            return find(elementNode, targetComponent, matcher);
        },
        findAll: (targetComponent, matcher) => {
            return findAll(elementNode, targetComponent, matcher);
        },
        findChild: (targetComponent, matcher) => {
            return findChild(elementNode, targetComponent, matcher);
        },
        findAllChildren: (targetComponent, matcher) => {
            return findAllChildren(elementNode, targetComponent, matcher);
        },
        maybeFind: (targetComponent, matcher) => {
            return maybeFind(elementNode, targetComponent, matcher);
        },
        maybeFindChild: (targetComponent, matcher) => {
            return maybeFindChild(elementNode, targetComponent, matcher);
        },
        toString: () => {
            return printNode(asRenderedNode(elementNode));
        },
        trigger: (eventPropName, eventArg) => {
            const eventFunction = elementNode.props[eventPropName];
            if (!eventFunction) {
                throw new MissingEventFunctionError({
                    componentName: name,
                    eventPropName,
                });
            }
            if (typeof eventFunction !== 'function') {
                throw new InvalidEventFunctionError({
                    componentName: name,
                    eventPropName,
                });
            }
            document.batchUpdate(() => {
                eventFunction(eventArg);
            });
        },
        isMatch(__component, __matcher) {
            return isMatch(asRenderedNode(elementNode), __component, __matcher);
        },
    };
    return asRenderedElementNode(elementNode);
};
