// @hubspot/ui-extensions/dist/testing/internal/fragment.js
import { RenderedNodeType } from "../types.js";
import { EMPTY_RENDERED_CHILD_NODES } from "./constants.js";
import { debugLog } from "./debug.js";
import { printNode } from "./print.js";
import { find, findAll, findAllChildren, findChild, maybeFind, maybeFindChild, } from "./query.js";
import { asRenderedFragmentNode, asRenderedNode, } from "./type-utils-internal.js";
/**
 * Creates a rendered fragment node.
 *
 * @param document The document that the fragment belongs to.
 * @returns A rendered fragment node.
 */
export const createFragmentNode = (document) => {
    const fragmentNode = {
        nodeType: RenderedNodeType.Fragment,
        text: null,
        childNodes: EMPTY_RENDERED_CHILD_NODES,
        document,
        debugLog: (label) => {
            debugLog(asRenderedNode(fragmentNode), label);
        },
        find: (targetComponent, matcher) => {
            return find(fragmentNode, targetComponent, matcher);
        },
        findAll: (targetComponent, matcher) => {
            return findAll(fragmentNode, targetComponent, matcher);
        },
        findChild: (targetComponent, matcher) => {
            return findChild(fragmentNode, targetComponent, matcher);
        },
        findAllChildren: (targetComponent, matcher) => {
            return findAllChildren(fragmentNode, targetComponent, matcher);
        },
        maybeFind: (targetComponent, matcher) => {
            return maybeFind(fragmentNode, targetComponent, matcher);
        },
        maybeFindChild: (targetComponent, matcher) => {
            return maybeFindChild(fragmentNode, targetComponent, matcher);
        },
        toString: () => {
            return printNode(asRenderedFragmentNode(fragmentNode));
        },
        isMatch(__component, __matcher) {
            return false;
        },
    };
    return asRenderedFragmentNode(fragmentNode);
};
