// @hubspot/ui-extensions/dist/testing/internal/root.js
import { RenderedNodeType } from "../types.js";
import { EMPTY_RENDERED_CHILD_NODES } from "./constants.js";
import { debugLog } from "./debug.js";
import { printNode } from "./print.js";
import { find, findAll, findAllChildren, findChild, maybeFind, maybeFindChild, } from "./query.js";
import { asRenderedNode } from "./type-utils-internal.js";
/**
 * Creates a rendered root node.
 *
 * @param document The document that the root belongs to.
 * @returns A rendered root node.
 */
export const createRootNode = (document) => {
    const rootNode = {
        nodeType: RenderedNodeType.Root,
        text: null,
        childNodes: EMPTY_RENDERED_CHILD_NODES,
        document,
        debugLog: (label) => {
            debugLog(asRenderedNode(rootNode), label);
        },
        find: (targetComponent, matcher) => {
            return find(rootNode, targetComponent, matcher);
        },
        findAll: (targetComponent, matcher) => {
            return findAll(rootNode, targetComponent, matcher);
        },
        findChild: (targetComponent, matcher) => {
            return findChild(rootNode, targetComponent, matcher);
        },
        findAllChildren: (targetComponent, matcher) => {
            return findAllChildren(rootNode, targetComponent, matcher);
        },
        maybeFind: (targetComponent, matcher) => {
            return maybeFind(rootNode, targetComponent, matcher);
        },
        maybeFindChild: (targetComponent, matcher) => {
            return maybeFindChild(rootNode, targetComponent, matcher);
        },
        toString: () => {
            return printNode(asRenderedNode(rootNode));
        },
        isMatch(__component, __matcher) {
            return false;
        },
    };
    return rootNode;
};
