// @hubspot/ui-extensions/dist/testing/internal/type-utils-internal.js
/**
 * Converts an internal rendered node to a public rendered node using a type assertion.
 *
 * @param node The node to convert.
 * @returns The converted node.
 */
export const asRenderedNode = (node) => {
    return node;
};
/**
 * Converts an internal rendered element node to a public rendered element node using a type assertion.
 *
 * @param node The node to convert.
 * @returns The converted node.
 */
export const asRenderedElementNode = (node) => {
    return node;
};
/**
 * Converts an internal rendered fragment node to a public rendered fragment node using a type assertion.
 *
 * @param node The node to convert.
 * @returns The converted node.
 */
export const asRenderedFragmentNode = (node) => {
    return node;
};
/**
 * Converts an internal rendered text node to a public rendered text node using a type assertion.
 *
 * @param node The node to convert.
 * @returns The converted node.
 */
export const asRenderedTextNode = (node) => {
    return node;
};
/**
 * Converts an internal rendered root node to a public rendered root node using a type assertion.
 *
 * @param node The node to convert.
 * @returns The converted node.
 */
export const asRenderedRootNode = (node) => {
    return node;
};
/**
 * Converts an internal rendered parent node to a public rendered parent node using a type assertion.
 *
 * @param node The node to convert.
 * @returns The converted node.
 */
export const asRenderedParentNode = (node) => {
    return node;
};
