// @hubspot/ui-extensions/dist/pages/internal/convert-page-routes-react-elements.js
import { Fragment, isValidElement, Children as ReactChildren, } from 'react';
import { PageRoutes, } from "../components/page-routes.js";
import { RouteNodeType, } from "./page-router-internal-types.js";
const isReactPageRoutesElement = (reactRouteNode) => {
    return isValidElement(reactRouteNode) && reactRouteNode.type === PageRoutes;
};
const isReactRouteElement = (reactNode) => isValidElement(reactNode) && reactNode.type === PageRoutes.Route;
const isReactIndexRouteElement = (reactNode) => isValidElement(reactNode) && reactNode.type === PageRoutes.IndexRoute;
const isReactAnyRouteElement = (reactNode) => isValidElement(reactNode) && reactNode.type === PageRoutes.AnyRoute;
const isReactFragmentElement = (reactRouteNode) => isValidElement(reactRouteNode) && reactRouteNode.type === Fragment;
/**
 * Describes a React node by producing a string representation of the React node.
 *
 * @param reactNode - The react node to describe.
 * @returns The description of the react node.
 */
const describeReactNode = (reactNode) => {
    if (reactNode == null) {
        return String(reactNode);
    }
    if (typeof reactNode !== 'object') {
        return JSON.stringify(reactNode);
    }
    if (isValidElement(reactNode)) {
        const elementType = reactNode.type;
        const name = typeof elementType === 'string'
            ? elementType
            : typeof elementType === 'function'
                ? elementType.displayName ||
                    elementType.name ||
                    'Unknown'
                : 'Unknown';
        return `<${name} />`;
    }
    return JSON.stringify(reactNode);
};
class InvalidRoutesReactNodeError extends Error {
    constructor(reactNode) {
        super(`Invalid React node for page routes: ${describeReactNode(reactNode)}. \nSee: https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensions/extension-points/app-pages/page-routing`);
    }
}
const addRoute = (parentNode, childNode) => {
    if (!parentNode.children) {
        parentNode.children = [];
    }
    parentNode.children.push(childNode);
};
/**
 * Converts a routes element to a routes node by recursively converting the routes children.
 *
 * @param reactRoutesElement - The routes element to convert.
 * @returns The converted routes node.
 */
function convertReactRoutesElement(reactRoutesElement) {
    const { path, layoutComponent, children } = reactRoutesElement.props;
    const routesNode = {
        type: RouteNodeType.Routes,
        path,
        layoutComponent,
        children: null,
    };
    convertChildrenReactRouteElements(routesNode, children);
    return routesNode;
}
/**
 * Converts the children of a routes element to a routes node by recursively converting the children.
 *
 * @param parentNode - The parent node to add the children to.
 * @param children - The children to convert.
 * @returns void.
 */
function convertChildrenReactRouteElements(parentNode, children) {
    ReactChildren.forEach(children, (child) => {
        if (child == null || typeof child === 'boolean') {
            return;
        }
        if (isReactPageRoutesElement(child)) {
            const routesNode = convertReactRoutesElement(child);
            addRoute(parentNode, routesNode);
        }
        else if (isReactRouteElement(child)) {
            const { path, component, id } = child.props;
            const routeNode = {
                type: RouteNodeType.Path,
                path,
                component,
                id,
            };
            addRoute(parentNode, routeNode);
        }
        else if (isReactIndexRouteElement(child)) {
            const { component, id } = child.props;
            const routeNode = {
                type: RouteNodeType.Path,
                path: '/',
                component,
                id,
            };
            addRoute(parentNode, routeNode);
        }
        else if (isReactAnyRouteElement(child)) {
            const { component, id } = child.props;
            const routeNode = {
                type: RouteNodeType.Path,
                path: '*',
                component,
                id,
            };
            addRoute(parentNode, routeNode);
        }
        else if (isReactFragmentElement(child)) {
            convertChildrenReactRouteElements(parentNode, child.props.children);
        }
        else {
            throw new InvalidRoutesReactNodeError(child);
        }
    });
}
/**
 * Converts a react node to a routes node by recursively converting the react node.
 *
 * @param reactNode - The react node to convert.
 * @returns The converted routes node.
 */
export const convertReactPageRoutesElement = (reactNode) => {
    if (isReactPageRoutesElement(reactNode)) {
        return convertReactRoutesElement(reactNode);
    }
    if (isReactFragmentElement(reactNode)) {
        const rootNode = {
            type: RouteNodeType.Routes,
            children: null,
        };
        convertChildrenReactRouteElements(rootNode, reactNode.props.children);
        return rootNode;
    }
    throw new InvalidRoutesReactNodeError(reactNode);
};
