// @hubspot/ui-extensions/dist/pages/create-page-router.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useMemo } from 'react';
import { EmptyState, Text } from "../shared/remoteComponents.js";
import { AppPageRouteProvider } from "./internal/app-page-route-context.js";
import { convertReactPageRoutesElement } from "./internal/convert-page-routes-react-elements.js";
import { RouteNodeType, } from "./internal/page-router-internal-types.js";
import { createTrieRouter } from "./internal/trie-router.js";
import { useAppPageLocation } from "./internal/useAppPageLocation.js";
/**
 * Adds the routes to the trie router.
 *
 * @param trieRouter - The trie router to add the routes to.
 * @param seenRouteIds - The set of route ids that have already been seen.
 * @param routeNode - The node to add the routes to.
 * @param parentPath - The path to the parent node.
 * @param parentLayouts - The layouts to apply to the node.
 */
const addRoutes = (trieRouter, seenRouteIds, routeNode, parentPath, parentLayouts) => {
    if (routeNode.type === RouteNodeType.Routes) {
        const prefix = parentPath + (routeNode.path ?? '');
        const layouts = routeNode.layoutComponent
            ? [...parentLayouts, routeNode.layoutComponent]
            : parentLayouts;
        if (routeNode.children) {
            for (const childRouteNode of routeNode.children) {
                addRoutes(trieRouter, seenRouteIds, childRouteNode /* routeNode */, prefix /* parentPath */, layouts /* parentLayouts */);
            }
        }
    }
    else {
        const fullPath = parentPath + routeNode.path;
        if (routeNode.id !== undefined) {
            if (seenRouteIds.has(routeNode.id)) {
                throw new Error(`Duplicate route id: "${routeNode.id}". \nSee: https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensions/extension-points/app-pages/page-routing#route-ids`);
            }
            seenRouteIds.add(routeNode.id);
        }
        trieRouter.addRoute(fullPath, {
            component: routeNode.component,
            layouts: parentLayouts,
            id: routeNode.id,
        });
    }
};
/**
 * The function type for [createPageRouter](https://developers.hubspot.com/docs/apps/developer-platform/add-features/ui-extensibility/app-pages/reference#createpagerouter).
 *
 * Example usage:
 *
 * ```tsx
 * const PageRouter = createPageRouter(
 *   <PageRoutes layoutComponent={AppLayout}>
 *     <PageRoutes.IndexRoute component={HomePage} id="home" />
 *     <PageRoutes.Route path="/docs" component={DocsPage} />
 *     <PageRoutes path="/customers" layoutComponent={CustomersLayout}>
 *       <PageRoutes.IndexRoute component={ListCustomersPage} />
 *       <PageRoutes.Route path="/:customerId" component={ViewCustomerPage} />
 *     </PageRoutes>
 *     <PageRoutes.AnyRoute component={NotFoundPage} />
 *   </PageRoutes>
 * );
 *
 * function AppPages() {
 *   return <PageRouter />;
 * }
 *
 * hubspot.extend<"pages">(() => <AppPages />);
 * ```
 *
 * @param routes The routes to render.
 * @returns The page router component.
 */
export const createPageRouter = (reactPageRoutesElement) => {
    // Validation errors (duplicate route ids, invalid wildcards, etc.) are captured
    // and deferred to render time instead of being thrown here. This is because
    // createPageRouter is typically called at module top level, before
    // hubspot.extend() registers the extension. If we throw here, the module
    // fails to evaluate and the extension host can only show a generic "unable to
    // initialize extension" message — the developer never sees the specific error.
    // By deferring the throw to the PageRouter component's render, the error
    // occurs inside the extension lifecycle where the host can display it.
    let initResult;
    try {
        // Convert the React element to a route node and add them to a trie router.
        const rootRouteNode = convertReactPageRoutesElement(reactPageRoutesElement);
        const trieRouter = createTrieRouter();
        const seenRouteIds = new Set();
        addRoutes(trieRouter, seenRouteIds, rootRouteNode /* routeNode */, '' /* parentPath */, [] /* parentLayouts */);
        initResult = { ok: true, trieRouter };
    }
    catch (e) {
        initResult = {
            ok: false,
            error: e instanceof Error ? e : new Error(String(e)),
        };
    }
    // Now create the PageRouter component that uses the trie router to match the current app page location.
    // The component will render the content of the matched route.
    // The component will also provide the matched app page route to the content of the matched route.
    const PageRouter = () => {
        if (!initResult.ok) {
            throw initResult.error;
        }
        const { trieRouter } = initResult;
        const appPageLocation = useAppPageLocation();
        const { path: appPagePath, params: appPageParams } = appPageLocation;
        const matchedRoute = useMemo(() => {
            return trieRouter.matchPath(appPagePath);
        }, [appPagePath]);
        const matchedAppPageRoute = useMemo(() => {
            return matchedRoute
                ? {
                    routeId: matchedRoute.data.id,
                    path: appPagePath,
                    params: {
                        ...matchedRoute.params,
                        ...appPageParams,
                    },
                }
                : {
                    path: appPagePath,
                    params: appPageParams,
                };
        }, [matchedRoute, appPagePath, appPageParams]);
        const routeData = matchedRoute?.data;
        const content = useMemo(() => {
            if (!routeData) {
                // TODO: Localize the title and body text.
                return (_jsx(EmptyState, { title: "Page not found", layout: "vertical", reverseOrder: true, children: _jsx(Text, { children: "This app page does not exist." }) }));
            }
            const Component = routeData.component;
            let node = _jsx(Component, {});
            // Wrap the route component with any ancestor layout components found in the route tree.
            for (let i = routeData.layouts.length - 1; i >= 0; i--) {
                const Layout = routeData.layouts[i];
                node = _jsx(Layout, { children: node });
            }
            return node;
        }, [routeData]);
        return (_jsx(AppPageRouteProvider, { pageRoute: matchedAppPageRoute, children: content }));
    };
    return PageRouter;
};
